---
title: "scRNA-seq Analysis Notebook"
author: "Ziyi Wang"
date: "2024-10-01"
output:
  html_notebook:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    self_contained: true
    css: ./HTML_source/style.css
---

# 00.Setup environment

## 00-00. Project folder

The prerequisite data and directories structure to reproduce the results of this analysis:

```
Project Folder/
+ HTML_source/                                                                                     # CSS files for this R notebook file
+ 2D_versatile_he/                                                                                 # Pre-trained StarDist AI model
- 240726_VisiumHD/                                                                                 # 10x Visium HD output folder
  - 240726_VisiumHD/RJ_7_23_24.tif                                                                 # 10x Visium HD high resolution image
  - 240726_VisiumHD/analysis/Pax9KO/outs/binned_outputs/square_002um/filtered_feature_bc_matrix.h5 # 10x Visium HD data in 2x2 um bin
- 00.10xVisiumHDAna.Rproj                                                                          # R projection file
- 00.NIH_10xVisiumHD.html                                                                          # This R notebook file
- 3M-february-2018.txt                                                                             # 10x barcode whitelist
- E12.5.csv                                                                                        # ROIs in E12.5
- E13.5.csv                                                                                        # ROIs in E13.5
```

## 00-01.R & Conda Environment

```{r 00-01.load R packages and Conda}
# This is a setup chunk
Sys.setenv(BASH_ENV="~/.bashrc")
knitr::opts_chunk$set(engine.opts = list(bash="--login"))
# Set enough physical memory for parallel computing
options(future.globals.maxSize= 256*1024^3)
WD <- getwd()
project <- "PAX9_KO_VisiumHD"
list.of.packages <- c("dplyr","tidyverse")
#Packages that does not install yet
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
if(length(new.packages)) {install.packages(new.packages)}
library(dplyr)
library(tidyverse)
condaPath<-reticulate::conda_list()$python%>%
            grep("r-miniconda/envs/stardist_env",.,value=TRUE)%>%
            str_remove(.,"bin\\/python$")%>%
            paste0(.,"lib")
Sys.setenv(LD_LIBRARY_PATH=paste0(condaPath,":", Sys.getenv("LD_LIBRARY_PATH")))
PythonPath<-reticulate::conda_list()$python%>%
            grep("r-miniconda/envs/stardist_env",.,value=TRUE)
Sys.setenv(RETICULATE_PYTHON=PythonPath)
list.of.packages <- c("reticulate")
#Packages that does not install yet
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
if(length(new.packages)) {install.packages(new.packages)}
library(reticulate)
use_condaenv("stardist_env", required = TRUE)
polychrome = c(
  "#5A5156", "#E4E1E3", "#F6222E", "#FE00FA", "#16FF32", "#3283FE",
  "#FEAF16", "#B00068", "#1CFFCE", "#90AD1C", "#2ED9FF", "#DEA0FD",
  "#AA0DFE", "#F8A19F", "#325A9B", "#C4451C", "#1C8356", "#85660D",
  "#B10DA1", "#FBE426", "#1CBE4F", "#FA0087", "#FC1CBF", "#F7E1A0",
  "#C075A6", "#782AB6", "#AAF400", "#BDCDFF", "#822E1C", "#B5EFB5",
  "#7ED7D1", "#1C7F93", "#D85FF7", "#683B79", "#66B0FF", "#3B00FB"
)
# Install & load dependent R packages
list.of.packages <- c(
  "R.utils","data.table","Seurat","tidyr","Matrix", "patchwork", "ggplotify", 
  "grid","gridExtra","rmarkdown","htmltools","plyr","plotly","htmlTable","magrittr","xml2", "hdf5r",
  "openxlsx", "ggplot2","RColorBrewer", "cowplot", "Polychrome", "htmlwidgets",
  "stringr", "hwriter", "writexl", "arrow", "magick"
)
list.of.packages.Bio <- c(
  "GSVA","survcomp","SingleCellExperiment", "ggtreeExtra", "AnnotationHub"
)
#Packages that does not install yet
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.packages.Bio <- list.of.packages.Bio[!(list.of.packages.Bio %in% installed.packages()[, "Package"])]
#install required packages
if (!requireNamespace("BiocManager", quietly = TRUE)){install.packages("BiocManager")}
if(length(new.packages.Bio)) {BiocManager::install(new.packages.Bio)}
if(length(new.packages)) {install.packages(new.packages)}
#Set GitHub Auth
#https://usethis.r-lib.org/articles/git-credentials.html
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
if (!require(biomaRt)) devtools::install_github("grimbough/biomaRt")
if (!require(SeuratWrappers)) devtools::install_github('satijalab/seurat-wrappers')
if (!require(viridis)) install.packages("viridis")
if (!require(readr)) install.packages("readr")
if (!require(phateR)) install.packages("phateR")
if (!require(sceasy)) devtools::install_github("cellgeni/sceasy")
if (!require(SeuratDisk)) remotes::install_github("mojaveazure/seurat-disk")
# Loading all packages
invisible(
  lapply(
    c(
      list.of.packages,list.of.packages.Bio, "phateR","readr","viridis","biomaRt","SeuratWrappers","SeuratDisk"
    ),
    library, character.only = TRUE
  )
)
```

## 00-02.Python Environment

Important Python Libraries

- `stardist` is a deep-learning library for star-convex object detection of cells/nuclei.
- `geopandas` is an extension of the pandas library for working with geospatial data.
- `squidpy` is a collection of tools that are built on scanpy and anndata for the visualization and analysis of spatial gene expression data. It is used in this guide mainly for the scanpy and anndata libraries.
- `fastparquet` is a Python library for efficient reading and writing of Apache Parquet files.

```{python 00-02.Python-1}
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import anndata
import geopandas as gpd
import scanpy as sc

from tifffile import imread, imwrite
from csbdeep.utils import normalize
from stardist.models import StarDist2D
from shapely.geometry import Polygon, Point
from scipy import sparse
from matplotlib.colors import ListedColormap

import pickle
def get_picklable_objects(namespace):
    picklable = {}
    for name, obj in namespace.items():
        try:
            pickle.dumps(obj)
            picklable[name] = obj
        except (pickle.PicklingError, TypeError):
            continue
    return picklable
```

Data Visualization Helper Functions: We define functions to help with plotting results in the next code block.

```{python 00-02.Python-2}
# General image plotting functions
def plot_mask_and_save_image(title, gdf, img, cmap, output_name=None, bbox=None):
    if bbox is not None:
        # Crop the image to the bounding box
        cropped_img = img[bbox[1]:bbox[3], bbox[0]:bbox[2]]
    else:
        cropped_img = img

    # Plot options
    fig, axes = plt.subplots(1, 2, figsize=(12, 6))

    # Plot the cropped image
    axes[0].imshow(cropped_img, cmap='gray', origin='lower')
    axes[0].set_title(title)
    axes[0].axis('off')

    # Create filtering polygon
    if bbox is not None:
        bbox_polygon = Polygon([(bbox[0], bbox[1]), (bbox[2], bbox[1]), (bbox[2], bbox[3]), (bbox[0], bbox[3])])
        # Filter for polygons in the box
        intersects_bbox = gdf['geometry'].intersects(bbox_polygon)
        filtered_gdf = gdf[intersects_bbox]
    else:
        filtered_gdf=gdf

    # Plot the filtered polygons on the second axis
    filtered_gdf.plot(cmap=cmap, ax=axes[1])
    axes[1].axis('off')
    axes[1].legend(loc='upper left', bbox_to_anchor=(1.05, 1))


    # Save the plot if output_name is provided
    if output_name is not None:
        plt.savefig(output_name, bbox_inches='tight')  # Use bbox_inches='tight' to include the legend
    else:
        plt.show()

def plot_gene_and_save_image(title, gdf, gene, img, adata, bbox=None, output_name=None):

    if bbox is not None:
        # Crop the image to the bounding box
        cropped_img = img[bbox[1]:bbox[3], bbox[0]:bbox[2]]
    else:
        cropped_img = img

    # Plot options
    fig, axes = plt.subplots(1, 2, figsize=(12, 6))

    # Plot the cropped image
    axes[0].imshow(cropped_img, cmap='gray', origin='lower')
    axes[0].set_title(title)
    axes[0].axis('off')

    # Create filtering polygon
    if bbox is not None:
        bbox_polygon = Polygon([(bbox[0], bbox[1]), (bbox[2], bbox[1]), (bbox[2], bbox[3]), (bbox[0], bbox[3])])


    # Find a gene of interest and merge with the geodataframe
    gene_expression = adata[:, gene].to_df()
    gene_expression['id'] = gene_expression.index
    merged_gdf = gdf.merge(gene_expression, left_on='id', right_on='id')

    if bbox is not None:
        # Filter for polygons in the box
        intersects_bbox = merged_gdf['geometry'].intersects(bbox_polygon)
        filtered_gdf = merged_gdf[intersects_bbox]
    else:
        filtered_gdf = merged_gdf

    # Plot the filtered polygons on the second axis
    filtered_gdf.plot(column=gene, cmap='inferno', legend=True, ax=axes[1])
    axes[1].set_title(gene)
    axes[1].axis('off')
    axes[1].legend(loc='upper left', bbox_to_anchor=(1.05, 1))

    # Save the plot if output_name is provided
    if output_name is not None:
        plt.savefig(output_name, bbox_inches='tight')  # Use bbox_inches='tight' to include the legend
    else:
        plt.show()

def plot_clusters_and_save_image(title, gdf, img, adata, bbox=None, color_by_obs=None, output_name=None, color_list=None):
    color_list=["#7f0000","#808000","#483d8b","#008000","#bc8f8f","#008b8b","#4682b4","#000080","#d2691e","#9acd32","#8fbc8f","#800080","#b03060","#ff4500","#ffa500","#ffff00","#00ff00","#8a2be2","#00ff7f","#dc143c","#00ffff","#0000ff","#ff00ff","#1e90ff","#f0e68c","#90ee90","#add8e6","#ff1493","#7b68ee","#ee82ee"]
    if bbox is not None:
        cropped_img = img[bbox[1]:bbox[3], bbox[0]:bbox[2]]
    else:
        cropped_img = img

    fig, axes = plt.subplots(1, 2, figsize=(12, 6))

    axes[0].imshow(cropped_img, cmap='gray', origin='lower')
    axes[0].set_title(title)
    axes[0].axis('off')

    if bbox is not None:
        bbox_polygon = Polygon([(bbox[0], bbox[1]), (bbox[2], bbox[1]), (bbox[2], bbox[3]), (bbox[0], bbox[3])])

    unique_values = adata.obs[color_by_obs].astype('category').cat.categories
    num_categories = len(unique_values)

    if color_list is not None and len(color_list) >= num_categories:
        custom_cmap = ListedColormap(color_list[:num_categories], name='custom_cmap')
    else:
        # Use default tab20 colors if color_list is insufficient
        tab20_colors = plt.cm.tab20.colors[:num_categories]
        custom_cmap = ListedColormap(tab20_colors, name='custom_tab20_cmap')

    merged_gdf = gdf.merge(adata.obs[color_by_obs].astype('category'), left_on='id', right_index=True)

    if bbox is not None:
        intersects_bbox = merged_gdf['geometry'].intersects(bbox_polygon)
        filtered_gdf = merged_gdf[intersects_bbox]
    else:
        filtered_gdf = merged_gdf

    # Plot the filtered polygons on the second axis
    plot = filtered_gdf.plot(column=color_by_obs, cmap=custom_cmap, ax=axes[1], legend=True)
    axes[1].set_title(color_by_obs)
    legend = axes[1].get_legend()
    legend.set_bbox_to_anchor((1.05, 1))
    axes[1].axis('off')

    # Move legend outside the plot
    plot.get_legend().set_bbox_to_anchor((1.25, 1))

    if output_name is not None:
        plt.savefig(output_name, bbox_inches='tight')
    else:
        plt.show()

# Plotting function for nuclei area distribution
def plot_nuclei_area(gdf,area_cut_off):
    fig, axs = plt.subplots(1, 2, figsize=(15, 4))
    # Plot the histograms
    axs[0].hist(gdf['area'], bins=50, edgecolor='black')
    axs[0].set_title('Nuclei Area')

    axs[1].hist(gdf[gdf['area'] < area_cut_off]['area'], bins=50, edgecolor='black')
    axs[1].set_title('Nuclei Area Filtered:'+str(area_cut_off))

    plt.tight_layout()
    plt.show()

# Total UMI distribution plotting function
def total_umi(adata_, cut_off):
    fig, axs = plt.subplots(1, 2, figsize=(12, 4))

    # Box plot
    axs[0].boxplot(adata_.obs["total_counts"], vert=False, widths=0.7, patch_artist=True, boxprops=dict(facecolor='skyblue'))
    axs[0].set_title('Total Counts')

    # Box plot after filtering
    axs[1].boxplot(adata_.obs["total_counts"][adata_.obs["total_counts"] > cut_off], vert=False, widths=0.7, patch_artist=True, boxprops=dict(facecolor='skyblue'))
    axs[1].set_title('Total Counts > ' + str(cut_off))

    # Remove y-axis ticks and labels
    for ax in axs:
        ax.get_yaxis().set_visible(False)

    plt.tight_layout()
    plt.show()
```

## 00-03.Online Resources

Retrieve internet resources for functional enrichment analysis (FEA).

```{r 00-03.GetOnlineResources}
## Get bioMart database list ----
SpeciesLs <- NULL
attempt <- 1
attempt_max <- 12
while (is.null(SpeciesLs) && attempt < 13) {
  if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
    url <- "https://www.ensembl.org"
  } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
    url <- "https://asia.ensembl.org"
  } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
    url <- "https://uswest.ensembl.org/"
  } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
    url <- "https://useast.ensembl.org/"
  }
  try({
    SpeciesLs <- useMart("ENSEMBL_MART_ENSEMBL", host = url) %>%
      biomaRt::listDatasets(mart = .) %>% as.data.frame()
  })
}
## Get KEGG organism list ----
KEGGSpeciesLs<-NULL
attemp<-1
while (is.null(KEGGSpeciesLs) && attemp<10) {
  url <- "https://rest.kegg.jp/list/organism"
  lines <- readLines(url)
  split <- strsplit(lines, "\t")
  u <- unlist(split)
  KEGGSpeciesLs <- matrix(u, ncol=4, byrow=TRUE)
  colnames(KEGGSpeciesLs) <-  c("T.number", "organism", "species", "phylogeny")
  KEGGSpeciesLs<-as.data.frame(KEGGSpeciesLs)
  attemp<-attemp+1
}
## Find overlapped between KEGG and BioMart ----
## as the organism list for our software
Species_BioMart<-SpeciesLs$description %>% 
  str_remove_all(., " genes \\(.*\\)$")
Species_KEGG<-KEGGSpeciesLs$species %>%
  str_extract_all(., "\\(.*\\)$")
KEGGSpeciesLs<-KEGGSpeciesLs[
  !(lapply(Species_KEGG, is_empty)%>%unlist()),
]
SpeciesLs<-cbind(
  SpeciesLs,SpeciesName=Species_BioMart%>%tolower()
)
KEGGSpeciesLs<-cbind(
  KEGGSpeciesLs,
  SpeciesName=Species_KEGG%>%unlist()%>%
    str_remove_all(.,"[\\(\\)]")%>%tolower()
)
KEGGSpeciesLs$SpeciesName[KEGGSpeciesLs$SpeciesName=="house mouse"]<-"mouse"
SpeLsIntersect<-intersect(SpeciesLs$SpeciesName, KEGGSpeciesLs$SpeciesName)
KEGGSpeciesLs<-KEGGSpeciesLs[
  KEGGSpeciesLs$SpeciesName%in%SpeLsIntersect,
]
SpeciesLs<-SpeciesLs[
  SpeciesLs$SpeciesName%in%SpeLsIntersect,
]
OrgList<-merge(KEGGSpeciesLs, SpeciesLs, by="SpeciesName")
rownames(OrgList)<-OrgList$description
SpeciesLs<-rownames(OrgList)
rm(KEGGSpeciesLs,SpeLsIntersect,Species_KEGG,Species_BioMart)
setAnnotationHubOption("CACHE", file.path(".", "02.Cache"))
OrgDb_ls <- BiocManager::available() %>% grep("^org\\..*\\.eg\\.db$", ., value = TRUE)
R.utils::setOption("clusterProfiler.download.method", "wget")
```

## 00-04.Additional environments

```{r 00-04}
# install dependencies
if(!require("hdf5r")){install.packages("hdf5r")}
if(!require("Seurat")){install.packages("Seurat")}
if(!require("SeuratObject")){install.packages("SeuratObject")}
if(!require("SeuratDisk")){remotes::install_github("mojaveazure/seurat-disk")}
# install platform specific source package
if (!require("loupeR")){
  os <- sub("Darwin", "macOS", Sys.info()["sysname"])
  url <- paste0("https://github.com/10XGenomics/loupeR/releases/latest/download/loupeR_", os, ".tar.gz")
  install.packages(url, repos = NULL, type = "source")
  loupeR::setup()
}
library(hdf5r)
library(Seurat)
library(SeuratObject)
library(SeuratDisk)
library(loupeR)
library(gtools)
```

## 00-05. Load functions

### Make function to list all 10x TSV files (scRNAseq, Visium, and Xenium).

```{r 00-05-1.List all 10x TSV files}
Find10xTSV <- function(WD){
  Path_Sep <- file.path("a","a") %>% str_remove_all("a")
  TSVmeta <- setNames(data.frame(matrix(ncol = 3, nrow = 0)), c("ID","Path", "Type"))
  fls <- list.files(
    path = WD, pattern = "barcodes.tsv.gz",
    full.names = TRUE, all.files = FALSE, recursive = TRUE
  )
  temp <- fls %>%
    .[!.[]%in%grep("raw_feature_bc_matrix",.,value = TRUE)]%>%
    .[!.[]%in%grep("SC_RNA_COUNTER_CS",.,value = TRUE)] %>%
    grep("/(count|outs|cell_feature_matrix)/", ., value = TRUE) %>%
    {
      c(
        .,
        fls %>% str_remove_all(., WD) %>% 
          str_remove_all(., "barcodes.tsv.gz$") %>%
          str_split(., "\\/") %>%
          {fls[which(unlist(lapply(., length)) == 3)]} %>% 
          lapply(., function(x){return(tail(x,2)[1])}) %>% 
          unlist()
      )
    }
  for (path in temp) {
    row <- c(path)
    path <- sub(WD,"",path) %>% dirname()
    while (dirname(path) != "/") {
      path <- dirname(path)
    }
    ID <- basename(path)
    path <- file.path(WD, basename(path))
    test <- list.files(
      path = path, pattern = "morphology.ome.tif|morphology_focus.ome.tif|morphology_mip.ome.tif",
      full.names = TRUE, all.files = FALSE, recursive = TRUE
    )
    Type <- NULL
    if(!is_empty(test)&is.null(Type)){
      Type <- "Xenium"
      test <-character(0)
    }else{
      test <- list.files(
        path = path, pattern = "aligned_fiducials.jpg|detected_tissue_image.jpg|scalefactors_json.json|tissue_hires_image.png|tissue_lowres_image.png|tissue_positions_list.csv",
        full.names = TRUE, all.files = FALSE, recursive = TRUE
      )
    }
    if(!is_empty(test)&is.null(Type)){
      Type <- "Visum"
      test <-character(0)
    }else{
      test <- "other"
    }
    if(!is_empty(test)&is.null(Type)){
      Type <- "scRNAseq"
      rm(test)
    }
    TSVmeta[which(temp%in%row[1]),] <- c(ID, row%>%str_remove_all(.,"barcodes.tsv.gz"), Type)
  }
  rownames(TSVmeta)<-TSVmeta$ID
  return(TSVmeta)
}
```

### Make functions to convert 10x TSV files into Seurat object.

```{r 00-05-2.TSV2Seurat Conversion}
Zenium_TSV2Seurat <- function(TSVmeta, WD){
  if(!file.exists(file.path(WD, "01.PlotOutput"))){
    dir.create(file.path(WD, "01.PlotOutput"))
  }
  outsum<-data.frame()
  ID_mapping<-data.frame(
    ensemblID=NULL,
    GeneSymbol=NULL,
    Type=NULL
  )
  SampleSlt<-TSVmeta$ID
  Step1_resume<-FALSE
  for (i in 1:length(SampleSlt)) {
    print(paste0(Sys.time(), ": Processing ", SampleSlt[i]))
    # Load the dataset
    temp<-read_tsv(gzfile(file.path(TSVmeta[SampleSlt[i],"Path"],"features.tsv.gz")),
                   col_names = c("ensemblID", "GeneSymbol","Type"))
    if(ncol(temp) < 2){
      temp <- cbind(
        temp, temp,
        type = "Gene Expression"
      )
      colnames(temp) <- c("ensemblID", "GeneSymbol", "Type")
      temp$ensemblID <- temp$ensemblID %>% str_remove(., "^.*_")
      temp$GeneSymbol <- temp$GeneSymbol %>% str_remove(., "^.*_")
      write_tsv(
        temp, file = file.path(TSVmeta[SampleSlt[i],"Path"],"features.tsv"),
        col_names = FALSE
      )
      gzip(file.path(TSVmeta[SampleSlt[i],"Path"],"features.tsv"), overwrite = TRUE)
    }else if(ncol(temp) < 3){
      temp <- cbind(
        temp,
        Type = "Gene Expression"
      )
      temp$ensemblID <- temp$ensemblID %>% str_remove(., "^.*_")
      temp$GeneSymbol <- temp$GeneSymbol %>% str_remove(., "^.*_")
      write_tsv(
        temp, file = file.path(TSVmeta[SampleSlt[i],"Path"],"features.tsv"),
        col_names = FALSE
      )
      gzip(file.path(TSVmeta[SampleSlt[i],"Path"],"features.tsv"), overwrite = TRUE)
    }
    ID_mapping<-rbind(
      ID_mapping,
      temp[!temp$ensemblID%in%ID_mapping$ensemblID,]
    )
    rm(temp)
    if( (!file.exists(paste0(".//Raw.",SampleSlt[i],".rds"))) | (!Step1_resume) ){
      print(paste0("Processing ", SampleSlt[i], ": Routine processing"))
      if(TSVmeta$Type[i]=="Xenium"){
        step <- "Before"
        xenium.obj <- LoadXenium(dirname(TSVmeta$Path[i]), fov = SampleSlt[i])
        xenium.obj <- subset(xenium.obj, subset = nCount_Xenium > 0)
        xenium.obj@meta.data$orig.ident <- factor(x = rep(SampleSlt[i], ncol(xenium.obj)), levels = SampleSlt[i])
        dims <- round(min(ncol(xenium.obj),nrow(xenium.obj))/3)
        xenium.obj <- SCTransform(xenium.obj, assay = "Xenium")
        # Identify the 10 most highly variable genes
        top10 <- head(VariableFeatures(xenium.obj), 10)
        # plot variable features with and without labels
        plot1 <- VariableFeaturePlot(xenium.obj)
        plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".03.",step,"DoubletRemove.VariableFeatures.png")),
          width = 1600, height = 900
        )
        print(plot1 + plot2)
        dev.off()
        xenium.obj <- RunPCA(xenium.obj, npcs = dims, features = rownames(xenium.obj))
        # Examine and visualize PCA results a few different ways
        print(xenium.obj[["pca"]], dims = 1:5, nfeatures = 5)
        plot<-VizDimLoadings(xenium.obj, dims = 1:2, reduction = "pca")
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".04.1.",step,"DoubletRemove.PCA_Res1.png")),
          width = 1600, height = 900
        )
        print(plot)
        dev.off()
        plot<-DimPlot(xenium.obj, reduction = "pca")
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".04.2.",step,"DoubletRemove.PCA_Res2.png")),
          width = 1600, height = 900
        )
        print(plot)
        dev.off()
        if(ncol(xenium.obj)<500){
          cells<-ncol(xenium.obj)
        }else{
          cells<-500
        }
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".04.3.",step,"DoubletRemove.PCA_Heatmap1.png")),
          width = 1600, height = 900
        )
        DimHeatmap(xenium.obj, dims = 1, cells = cells, balanced = TRUE)
        dev.off()
        if(ncol(xenium.obj)<15){
          dimsupper<-ncol(xenium.obj)-1
        }else{
          dimsupper<-15
        }
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".04.4.",step,"DoubletRemove.PCA_Heatmap2.png")),
          width = 1600, height = 900
        )
        DimHeatmap(xenium.obj, dims = 1:dimsupper, cells = cells, balanced = TRUE)
        dev.off()
        Cutdim <- xenium.obj@reductions[["pca"]]@stdev%>%
          .[.>mean(.)]%>%length()
        if(ncol(xenium.obj)<50){
          plot2<-ElbowPlot(xenium.obj, ndims = ncol(xenium.obj)-1)
        }else{
          plot2<-ElbowPlot(xenium.obj, ndims = 50)
        }
        png(
          filename = file.path(
            WD, "01.PlotOutput", paste0(SampleSlt[i],".05.",step,"DoubletRemove.JackStraw_Elbow_cutDim_",Cutdim,"_.png")
          ),
          width = 1600, height = 900
        )
        print(plot2)
        dev.off()
        xenium.obj <- RunUMAP(xenium.obj, dims = 1:Cutdim)
        xenium.obj <- FindNeighbors(xenium.obj, reduction = "pca", dims = 1:Cutdim)
        xenium.obj <- FindClusters(xenium.obj, resolution = 0.3)
        print( paste0("Processing ", SampleSlt[i], ": Making Plots"))
        plot<-DimPlot(xenium.obj, reduction = "umap", cols = "polychrome")
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".06.",step,"DoubletRemove.Cluster.png")),
          width = 1600, height = 900
        )
        print(plot)
        dev.off()
        # Visualize QC metrics as a violin plot
        plot<-VlnPlot(xenium.obj, features = c("nFeature_Xenium", "nCount_Xenium"), ncol = 2)
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".01.",step,"DoubletRemove.FeatureVolin.split.png")),
          width = 1600, height = 900
        )
        print(plot)
        dev.off()
        plot<-VlnPlot(xenium.obj, features = c("nFeature_Xenium", "nCount_Xenium"), ncol = 2, group.by = "orig.ident")
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".01.",step,"DoubletRemove.FeatureVolin.stack.png")),
          width = 1600, height = 900
        )
        print(plot)
        dev.off()
        # FeatureScatter is typically used to visualize feature-feature relationships, but can be used
        # for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
        plot1 <- FeatureScatter(xenium.obj, feature1 = "nCount_Xenium", feature2 = "nFeature_Xenium")
        png(
          filename = file.path(WD, "01.PlotOutput", paste0(SampleSlt[i],".02.",step,"DoubletRemove.FeatureScatter.png")),
          width = 1600, height = 900
        )
        print(plot1)
        dev.off()
        graphics.off()
        gc()
        graphics.off()
        p0<-ImageDimPlot(
          xenium.obj, group.by = "seurat_clusters", size = 0.4, cols = "polychrome", dark.background = F
        ) + ggtitle("Seurat Clusters")
        p1<-DimPlot(xenium.obj, cols = "polychrome")
        p2<-FeaturePlot(xenium.obj, pt.size = 1, features = "nFeature_Xenium", slot = "data", min.cutoff = NA, max.cutoff = NA, keep.scale=NULL)+
          labs(title = paste0("# of Gene of each cell in ",SampleSlt[i]))
        p3<-FeaturePlot(xenium.obj, pt.size = 1, features = "nCount_Xenium", slot = "data", min.cutoff = NA, max.cutoff = NA, keep.scale=NULL)+ 
          labs(title = paste0("# of Reads of each cell in ",SampleSlt[i]))
        png(
          filename = file.path(
            WD, "01.PlotOutput", paste0(SampleSlt[i], ".00.BeforeDoubletRemove.DimPlot.png")
          ),
          width = 1200, height = 1000
        )
        grid.arrange(p0, p1, p2, p3, nrow=2, top = paste0("Feature Distribution in: ",SampleSlt[i]))
        dev.off()
        graphics.off()
        print(paste0("Processing ", SampleSlt[i], ": Saving"))
        saveRDS(
          xenium.obj, file = file.path(
            WD,
            paste0("Raw.", SampleSlt[i], ".rds")
          )
        )
        rm(xenium.obj)
        gc()
      }else{
        print(paste0(Sys.time(), ": ", SampleSlt[i]), " is not Xenium Data.")
      }
    }else{
      print(paste0(Sys.time(), ": Skip ", SampleSlt[i], "- Found previous results"))
    }
  }
  save(outsum, ID_mapping, file=file.path(WD, "Parameters.RData"))
}
```

### Make functions for gene ID conversion and functional enrichment analysis (FEA).

```{r 00-05-4.IDConvert_FEA_Functions}
Symbol2Ensembl_BioMat <- function(dataset = "mmusculus_gene_ensembl", GeneName){
  ensembl_gene_id <- NULL
  attempt <- 1
  attempt_max <- 12
  while (is.null(ensembl_gene_id) && attempt < 13) {
    if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
      url <- "https://www.ensembl.org"
    } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
      url <- "https://asia.ensembl.org"
    } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
      url <- "https://uswest.ensembl.org/"
    } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
      url <- "https://useast.ensembl.org/"
    }
    try({
      ensembl_gene_id <- getBM(
        mart = useDataset(dataset,useMart("ENSEMBL_MART_ENSEMBL", host = url)),
        attributes = c("ensembl_gene_id","entrezgene_id","gene_biotype",
                       "external_gene_name","description"),
        filter = "external_gene_name",
        values = GeneName,
        uniqueRows = TRUE)
    })
    attempt <- attempt + 1
    Sys.sleep(0.25)
  }
  return(ensembl_gene_id)
}
Entrez2Ensembl_BioMat <- function(dataset = "mmusculus_gene_ensembl", entrezgene_id){
  ensembl_gene_id <- NULL
  attempt <- 1
  attempt_max <- 12
  while (is.null(ensembl_gene_id) && attempt < 13) {
    if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
      url <- "https://www.ensembl.org"
    } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
      url <- "https://asia.ensembl.org"
    } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
      url <- "https://uswest.ensembl.org/"
    } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
      url <- "https://useast.ensembl.org/"
    }
    try({
      ensembl_gene_id <- getBM(
        mart = useDataset(dataset,useMart("ENSEMBL_MART_ENSEMBL", host = url)),
        attributes = c("ensembl_gene_id","entrezgene_id","gene_biotype",
                       "external_gene_name","description"),
        filter = "entrezgene_id",
        values = entrezgene_id,
        uniqueRows = TRUE)
    })
    attempt <- attempt + 1
    Sys.sleep(0.25)
  }
  return(ensembl_gene_id)
}
RetrieveEnsembl_BioMat <- function(dataset = "mmusculus_gene_ensembl", ensembl_gene_id){
  RetrieveEnsembl <- NULL
  attempt <- 1
  attempt_max <- 12
  while (is.null(RetrieveEnsembl) && attempt < 13) {
    if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
      url <- "https://www.ensembl.org"
    } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
      url <- "https://asia.ensembl.org"
    } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
      url <- "https://uswest.ensembl.org/"
    } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
      url <- "https://useast.ensembl.org/"
    }
    try({
      RetrieveEnsembl <- getBM(
        mart = useDataset(dataset,useMart("ENSEMBL_MART_ENSEMBL", host = url)),
        attributes = c("ensembl_gene_id","entrezgene_id","gene_biotype",
                       "external_gene_name","description"),
        filter = "ensembl_gene_id",
        values = ensembl_gene_id,
        uniqueRows = TRUE)
    })
    attempt <- attempt + 1
    Sys.sleep(0.25)
  }
  return(RetrieveEnsembl)
}
RetrieveAnno_BioMat <- function(dataset = "mmusculus_gene_ensembl", genels){
  ensembl_gene_id <- genels[grepl("^ENS.*[0-9]{11}$",genels)]
  if(!is_empty(ensembl_gene_id)){
    ensembl_gene_id <- RetrieveEnsembl_BioMat(dataset = dataset, ensembl_gene_id = ensembl_gene_id)
  }
  entrezgene_id <- genels[grepl("^[0-9]$",genels)]
  if(!is_empty(entrezgene_id)){
    entrezgene_id <- Entrez2Ensembl_BioMat(dataset = dataset, entrezgene_id = entrezgene_id)
  }
  external_gene_name <- genels[!genels%in%c(ensembl_gene_id, entrezgene_id)]
  if(!is_empty(external_gene_name)){
    external_gene_name <- Symbol2Ensembl_BioMat(dataset = dataset, GeneName = external_gene_name)
  }
  Anno_gene <- rbind(
    ensembl_gene_id,
    entrezgene_id,
    external_gene_name
  )
  return(Anno_gene)
}
GetGeneAnno_Pubmed <- function(x, Par_species.Name, Par_species.Taxid, dataset){
  Par_cl<-makeCluster(parallelly::freeConnections()-3)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=0, max=length(x), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  ## Refer here for all column names for
  ## Entrez API (https://www.ncbi.nlm.nih.gov/books/NBK25501/)
  temp2<-foreach(
    i=1:length(x), .errorhandling = "remove",
    .inorder=FALSE, .options.snow=Par_opts,
    .packages = c(
      "dplyr","utils","xml2","S4Vectors", "data.table",
      "stringr","BiocGenerics","rlang", "biomaRt"
    ),
    .noexport = c(
      "out", "attemp", "NCBIcheck", "entrezgene_id", "DataTable", "entrezgene_id",
      "external_gene_name", "description", "gene_biotype", "ensembl_gene_id",
      "RetrieveEnsembl", "attempt_max"
    )
  ) %dopar% {
    out <- NULL
    attemp <- 0
    while (is.null(out) | (attemp < 3)) {
      try({
        NCBIcheck <- numeric(0)
        NCBIcheck<-read_html(
          paste0(
            "https://www.ncbi.nlm.nih.gov/gene/?term=",
            x[i]
          )
        )
        entrezgene_id <- NCBIcheck %>%
          xml_find_all(., ".//span")%>%xml_text()%>%
          grep("Gene ID",.,value = TRUE)
        if(is_empty(entrezgene_id)){
          try({
            NCBIcheck<-read_html(
              paste0(
                "https://www.ncbi.nlm.nih.gov/gene/?term=",
                paste0('(',x[i],')AND"',Par_species.Name,'"[porgn:__txid',Par_species.Taxid,']') %>% 
                  URLencode()
              )
            )
            entrezgene_id <- NCBIcheck %>%
              xml_find_all(., ".//span")%>%xml_text()%>%
              grep("Gene ID",.,value = TRUE)
          })
        }
        if(is_empty(entrezgene_id)){
          NCBIcheck <- NCBIcheck %>% xml_find_all(., ".//span") %>% xml_text()%>%
            grep("ID\\: [0-9].*$",.,value = TRUE) %>%
            str_remove_all(.,"^ID\\: ") %>% .[1] %>%
            {read_html(paste0("https://www.ncbi.nlm.nih.gov/gene/?term=", .))}
          entrezgene_id <- NCBIcheck %>%
            xml_find_all(., ".//span")%>%xml_text()%>%
            grep("Gene ID",.,value = TRUE) 
        }
        if(!is_empty(entrezgene_id)){
          DataTable <- NCBIcheck %>% xml_find_all(., ".//*[self::dt or self::dd]") %>% xml_text()
          entrezgene_id <- entrezgene_id %>% unlist() %>% str_extract_all(.,"(\\d)+")%>%.[[1]]%>%.[1]
          external_gene_name <- DataTable[grep("[Ss]ymbol",DataTable)[1]+1] %>% str_split(.," ") %>% {unlist(.)[1]}%>%
            str_remove_all(.,"provided$")
          description <- DataTable[grep("(description|Full Name)",DataTable)[1]+1]
          gene_biotype <- DataTable[grep("Gene type",DataTable)[1]+1] %>% str_replace_all(., " ", "_")
          ensembl_gene_id <- DataTable[grep("See related",DataTable)[1]+1] %>%
            str_extract_all(., "ENS[A-Z]*[0-9]{11}") %>% unlist()
          if(grepl("^ENS.*[0-9]{11}$", x[i])){
            ensembl_gene_id <- x[i]
          }
          if(length(ensembl_gene_id)>1){
            RetrieveEnsembl <- NULL
            attempt <- 1
            attempt_max <- 12
            while (is.null(RetrieveEnsembl) && attempt < 13) {
              if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
                url <- "https://www.ensembl.org"
              } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
                url <- "https://asia.ensembl.org"
              } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
                url <- "https://uswest.ensembl.org/"
              } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
                url <- "https://useast.ensembl.org/"
              }
              try({
                RetrieveEnsembl <- getBM(
                  mart = useDataset(dataset,useMart("ENSEMBL_MART_ENSEMBL", host = url)),
                  attributes = c("ensembl_gene_id","entrezgene_id","gene_biotype",
                                 "external_gene_name","description"),
                  filter = "ensembl_gene_id",
                  values = ensembl_gene_id,
                  uniqueRows = TRUE)
              })
              attempt <- attempt + 1
              Sys.sleep(0.25)
            }
            if(external_gene_name%in%RetrieveEnsembl$external_gene_name){
              ensembl_gene_id <- RetrieveEnsembl$ensembl_gene_id[RetrieveEnsembl$external_gene_name%in%external_gene_name]
              ensembl_gene_id <- ensembl_gene_id[1]
              gene_biotype <- RetrieveEnsembl$gene_biotype[RetrieveEnsembl$external_gene_name%in%external_gene_name]
              gene_biotype <- gene_biotype[1]
            }else{
              ensembl_gene_id <- ensembl_gene_id[1]
            }
          }
        }else{
          ensembl_gene_id <- NA
          entrezgene_id <- NA
          gene_biotype <- NA
          external_gene_name <- NA
          description <- NA
        }
        out<-data.table(
          Query = x[i],
          ensembl_gene_id = ensembl_gene_id,
          entrezgene_id = entrezgene_id,
          gene_biotype = gene_biotype,
          external_gene_name = external_gene_name,
          description = description
        )
      })
      attemp <- attemp + 1
      Sys.sleep(0.25)
    }
    if(is.null(out)){
      out<-data.table(
        Query = x[i],
        ensembl_gene_id = NA,
        entrezgene_id = NA,
        gene_biotype = NA,
        external_gene_name = NA,
        description = NA
      )
    }
    return(out)
  }
  stopCluster(Par_cl)
  StatusOK <- lapply(temp2, function(x){
    out <- dim(x)[2]
    if(is.null(out)){
      out <- 0
    }
    return(out)
  }) %>% unlist()
  return(rbindlist(temp2[StatusOK>0]))
}
GetEntrezID_Pubmed <- function(x, Par_species.Name, Par_species.Taxid, dataset){
  Par_cl<-makeCluster(parallelly::freeConnections()-3)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=0, max=length(x), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  ## Refer here for all column names for
  ## Entrez API (https://www.ncbi.nlm.nih.gov/books/NBK25501/)
  temp2<-foreach(
    i=1:length(x), .errorhandling = "remove",
    .inorder=FALSE, .options.snow=Par_opts,
    .packages = c(
      "dplyr","utils","xml2","S4Vectors", "data.table",
      "stringr","BiocGenerics","rlang", "biomaRt"
    ),
    .noexport = c(
      "out", "attemp", "NCBIcheck", "entrezgene_id", "DataTable"
    )
  ) %dopar% {
    out <- NULL
    attemp <- 0
    while (is.null(out) | (attemp < 3)) {
      try({
        NCBIcheck <- numeric(0)
        NCBIcheck<-read_html(
          paste0(
            "https://www.ncbi.nlm.nih.gov/gene/?term=",
            x[i]
          )
        )
        entrezgene_id <- NCBIcheck %>%
          xml_find_all(., ".//span")%>%xml_text()%>%
          grep("Gene ID",.,value = TRUE)
        if(is_empty(entrezgene_id)){
          try({
            NCBIcheck<-read_html(
              paste0(
                "https://www.ncbi.nlm.nih.gov/gene/?term=",
                paste0('(',x[i],')AND"',Par_species.Name,'"[porgn:__txid',Par_species.Taxid,']') %>% 
                  URLencode()
              )
            )
            entrezgene_id <- NCBIcheck %>%
              xml_find_all(., ".//span")%>%xml_text()%>%
              grep("Gene ID",.,value = TRUE)
          })
        }
        if(is_empty(entrezgene_id)){
          NCBIcheck <- NCBIcheck %>% xml_find_all(., ".//span") %>% xml_text()%>%
            grep("ID\\: [0-9].*$",.,value = TRUE) %>%
            str_remove_all(.,"^ID\\: ") %>% .[1] %>%
            {read_html(paste0("https://www.ncbi.nlm.nih.gov/gene/?term=", .))}
          entrezgene_id <- NCBIcheck %>%
            xml_find_all(., ".//span")%>%xml_text()%>%
            grep("Gene ID",.,value = TRUE) 
        }
        if(!is_empty(entrezgene_id)){
          DataTable <- NCBIcheck %>% xml_find_all(., ".//*[self::dt or self::dd]") %>% xml_text()
          entrezgene_id <- entrezgene_id %>% unlist() %>% str_extract_all(.,"(\\d)+")%>%.[[1]]%>%.[1]
        }else{
          entrezgene_id <- NA
        }
        out<-data.table(
          Query = x[i],
          entrezgene_id = entrezgene_id
        )
      })
      attemp <- attemp + 1
      Sys.sleep(0.25)
    }
    if(is.null(out)){
      out<-data.table(
        Query = x[i],
        entrezgene_id = NA
      )
    }
    return(out)
  }
  stopCluster(Par_cl)
  StatusOK <- lapply(temp2, function(x){
    out <- dim(x)[2]
    if(is.null(out)){
      out <- 0
    }
    return(out)
  }) %>% unlist()
  return(rbindlist(temp2[StatusOK>0]))
}
GetGeneAnno_UniProt <- function(Query, Par_species.Name, Par_species.Taxid, dataset){
  Par_cl<-makeCluster(parallelly::freeConnections()-3)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=0, max=length(Query), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  UniProtAnno<-foreach(
    i=1:length(Query), .errorhandling = "remove",
    .inorder=FALSE, .options.snow=Par_opts,
    .packages = c("dplyr","utils","xml2","S4Vectors", "data.table","stringr"),
    .noexport = c("temp", "entrezgene_id", "gene_biotype", "external_gene_name", "ensembl_gene_id")
  ) %dopar% {
    gene_biotype <- "protein_coding"
    temp <- data.table::fread(
      paste0(
        "https://rest.uniprot.org/uniprotkb/search?query=",
        Query[i],
        "+organism_id:",Par_species.Taxid,
        "&format=tsv&fields=xref_ensembl,xref_geneid,gene_primary,protein_name"
      ),verbose=FALSE,showProgress=TRUE
    )[1,]
    if((sum(is.na(temp))==4)&grepl("\\.[0-9]$", Query[i])){
      temp <- data.table::fread(
        paste0(
          "https://rest.uniprot.org/uniprotkb/search?query=",
          str_remove_all(Query[i], "\\.[0-9]$"),
          "+organism_id:",Par_species.Taxid,
          "&format=tsv&fields=xref_ensembl,xref_geneid,gene_primary,protein_name"
        ),verbose=FALSE,showProgress=TRUE
      )[1,]
    }
    if(sum(is.na(temp))==4){
      gene_biotype <- NA
    }
    colnames(temp) <- c("ensembl_gene_id", "entrezgene_id", "external_gene_name", "description")
    ensembl_gene_id = temp$ensembl_gene_id %>% str_remove_all(.,"\\.[0-9];$")
    if(grepl("^ENS[A-Z]*[0-9]{11}$", Query[i])){
      ensembl_gene_id = Query[i]
    }
    ensembl_gene_id <- ensembl_gene_id %>% str_split(.,";") %>% unlist() %>% str_remove_all(., "\\.[0-9]$")
    entrezgene_id <- temp$entrezgene_id %>% str_remove_all(.,";$")
    if(length(ensembl_gene_id)>1){
      external_gene_name <- temp$external_gene_name
      RetrieveEnsembl <- NULL
      attempt <- 1
      attempt_max <- 12
      while (is.null(RetrieveEnsembl) && attempt < 13) {
        if (attempt %in% seq(from = 1, to = attempt_max, 3)) {
          url <- "https://www.ensembl.org"
        } else if (attempt %in% seq(from = 2, to = attempt_max, 3)) {
          url <- "https://asia.ensembl.org"
        } else if (attempt %in% seq(from = 3, to = attempt_max, 3)) {
          url <- "https://uswest.ensembl.org/"
        } else if (attempt %in% seq(from = 4, to = attempt_max, 3)) {
          url <- "https://useast.ensembl.org/"
        }
        try({
          RetrieveEnsembl <- getBM(
            mart = useDataset(dataset,useMart("ENSEMBL_MART_ENSEMBL", host = url)),
            attributes = c("ensembl_gene_id","entrezgene_id","gene_biotype",
                           "external_gene_name","description"),
            filter = "ensembl_transcript_id",
            values = ensembl_gene_id,
            uniqueRows = TRUE)
        })
        attempt <- attempt + 1
        Sys.sleep(0.25)
      }
      if(external_gene_name%in%RetrieveEnsembl$external_gene_name){
        ensembl_gene_id <- RetrieveEnsembl$ensembl_gene_id[RetrieveEnsembl$external_gene_name%in%external_gene_name]
        ensembl_gene_id <- ensembl_gene_id[1]
        gene_biotype <- RetrieveEnsembl$gene_biotype[RetrieveEnsembl$external_gene_name%in%external_gene_name]
        gene_biotype <- gene_biotype[1]
        if(is.na(temp$entrezgene_id)){
          entrezgene_id <- RetrieveEnsembl$entrezgene_id[RetrieveEnsembl$external_gene_name%in%external_gene_name]
          entrezgene_id <- entrezgene_id[1]
        }
      }else{
        ensembl_gene_id <- ensembl_gene_id[1]
      }
    }
    data.table(
      GeneSymbol = Query[i],
      ensembl_gene_id = ensembl_gene_id,
      entrezgene_id = entrezgene_id,
      gene_biotype = gene_biotype,
      external_gene_name = temp$external_gene_name,
      description = temp$description
    ) %>% return()
  }
  stopCluster(Par_cl)
  return(rbindlist(UniProtAnno))
}
GeneName2ID_PubMedBioMat_UniProt <- function(
    genels, OrgList, species, dataset = OrgList[species, "dataset"], ShinyProgress = FALSE, ShinyProg.N = 1
){
  steps <- 7
  library(biomaRt)
  library(parallel)
  library(foreach)
  library(doSNOW)
  library(mygene)
  library(data.table)
  library(xml2)
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Gene ID Convert by BioMart")}
  Anno_gene <- RetrieveAnno_BioMat(
    dataset = dataset,
    genels = genels %>% str_remove_all(., "\\.[0-9]$") %>% unique()
  )
  Anno_gene <- Anno_gene[order(Anno_gene$ensembl_gene_id),]
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "EntrezID ID Convert by PubMed")}
  Anno<-Anno_gene[!is.na(Anno_gene$entrezgene_id),] #For EntrezGeneID
  Anno<-Anno[!duplicated(Anno[,'ensembl_gene_id']),] #For EntrezGeneID
  Par_species.Taxid <- getGenes(Anno$entrezgene_id[sample(1:nrow(Anno))[1:20]])$taxid %>% 
    unique() %>% {.[!is.na(.[])]}
  Par_species.Name <- OrgList[species, "species"] %>% str_remove_all(., " \\(.*\\)$")
  rm(Anno)
  gc()
  ensembl_gene_id <- Anno_gene$ensembl_gene_id[Anno_gene$external_gene_name%in%""]
  if(!is_empty(ensembl_gene_id)){
    PubMedEntrezRes <- GetGeneAnno_Pubmed(
      ensembl_gene_id, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    Query <- 1:length(PubMedEntrezRes$Query)
    names(Query) <- PubMedEntrezRes$Query
    PubMedEntrezRes <- as.data.frame(PubMedEntrezRes)
    PubMedEntrezRes <- PubMedEntrezRes[Query[Anno_gene$ensembl_gene_id[Anno_gene$external_gene_name==""]],]
    Anno_gene$external_gene_name[Anno_gene$external_gene_name%in%""] <- PubMedEntrezRes$external_gene_name
    Anno_gene$description[Anno_gene$external_gene_name%in%""] <- PubMedEntrezRes$description
    rm(ensembl_gene_id, Query, PubMedEntrezRes)
    gc()
  }
  ensembl_gene_id <- Anno_gene$ensembl_gene_id[is.na(Anno_gene$entrezgene_id)]
  if(!is_empty(ensembl_gene_id)){
    PubMedEntrezRes <- GetEntrezID_Pubmed(
      ensembl_gene_id, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    Query <- 1:length(PubMedEntrezRes$Query)
    names(Query) <- PubMedEntrezRes$Query
    PubMedEntrezRes <- as.data.frame(PubMedEntrezRes)
    PubMedEntrezRes <- PubMedEntrezRes[Query[Anno_gene$ensembl_gene_id[is.na(Anno_gene$entrezgene_id)]],]
    Anno_gene$entrezgene_id[is.na(Anno_gene$entrezgene_id)] <- PubMedEntrezRes$entrezgene_id
    rm(ensembl_gene_id, Query, PubMedEntrezRes)
    gc()
  }
  entrezgene_id.na <- Anno_gene$entrezgene_id[!is.na(Anno_gene$entrezgene_id)&is.na(Anno_gene$external_gene_name)]
  entrezgene_id <- Anno_gene$entrezgene_id[Anno_gene$entrezgene_id %in% entrezgene_id.na]
  if(!is_empty(entrezgene_id)){
    PubMedEntrezRes <- GetGeneAnno_Pubmed(
      entrezgene_id, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    #
    external_gene_name <- PubMedEntrezRes$external_gene_name
    names(external_gene_name) <- PubMedEntrezRes$Query
    OrigSltIdx <- Anno_gene$entrezgene_id %in% names(external_gene_name)
    SltRowIdx <- Anno_gene$entrezgene_id[OrigSltIdx]
    Anno_gene$external_gene_name[OrigSltIdx] <- external_gene_name[SltRowIdx] %>% as.character()
    #
    description <- PubMedEntrezRes$description
    names(description) <- PubMedEntrezRes$Query
    OrigSltIdx <- Anno_gene$entrezgene_id %in% names(description)
    SltRowIdx <- Anno_gene$entrezgene_id[OrigSltIdx]
    Anno_gene$description[OrigSltIdx] <- description[SltRowIdx] %>% as.character()
    rm(OrigSltIdx, SltRowIdx, entrezgene_id.na, entrezgene_id, external_gene_name, description, PubMedEntrezRes)
    gc()
  }
  external_gene_name.dup <- Anno_gene$external_gene_name[duplicated(Anno_gene$external_gene_name, incomparables = NA)]
  OrigSltIdx <- (Anno_gene$external_gene_name %in% external_gene_name.dup) & !is.na(Anno_gene$entrezgene_id)
  entrezgene_id <- Anno_gene$entrezgene_id[OrigSltIdx]
  if(!is_empty(entrezgene_id)){
    PubMedEntrezRes <- GetGeneAnno_Pubmed(
      entrezgene_id, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    SltRowIdx <- Anno_gene$entrezgene_id[OrigSltIdx]
    #
    external_gene_name <- PubMedEntrezRes$external_gene_name
    names(external_gene_name) <- PubMedEntrezRes$Query
    Anno_gene$external_gene_name[OrigSltIdx] <- external_gene_name[SltRowIdx] %>% as.character()
    #
    description <- PubMedEntrezRes$description
    names(description) <- PubMedEntrezRes$Query
    Anno_gene$description[OrigSltIdx] <- description[SltRowIdx] %>% as.character()
    rm(external_gene_name.dup, OrigSltIdx, SltRowIdx, entrezgene_id, external_gene_name, description, PubMedEntrezRes)
    gc()
  }
  Anno_gene$external_gene_name[!is.na(Anno_gene$external_gene_name)] <- make.unique(
    Anno_gene$external_gene_name[!is.na(Anno_gene$external_gene_name)]
  )
  Anno_gene <- cbind(
    UniqueID = paste0(
      Anno_gene$ensembl_gene_id,"-",
      Anno_gene$entrezgene_id,"-",
      Anno_gene$external_gene_name
    ), Anno_gene
  )
  Anno_gene <- Anno_gene[!duplicated(Anno_gene$UniqueID),]
  rownames(Anno_gene) <- Anno_gene$UniqueID
  Par_cl<-makeCluster(parallelly::freeConnections()-3)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=0, max=length(genels), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  ID_mapping<-foreach(
    i=1:length(genels), .errorhandling = "remove",
    .inorder=FALSE, .options.snow=Par_opts,
    .packages = c("dplyr","utils","xml2","S4Vectors", "data.table","stringr"),
    .noexport = c("x", "temp")
  ) %dopar% {
    x <- genels[i]
    temp <- Anno_gene[Anno_gene$external_gene_name%in%x,]
    if(nrow(temp)<1){temp <- Anno_gene[Anno_gene$ensembl_gene_id%in%x,]}
    if(nrow(temp)<1){temp <- Anno_gene[Anno_gene$entrezgene_id%in%x,]}
    if((nrow(temp)<1)&grepl("\\.[0-9]$", x)){
      temp <- Anno_gene[Anno_gene$external_gene_name%in%str_remove_all(x,"\\.[0-9]$"),]
    }
    if(nrow(temp)){
      temp <- cbind(Query = genels[i], temp)
      return(data.table(temp))
    }
  }
  stopCluster(Par_cl)
  ID_mapping <- rbindlist(ID_mapping)
  ID_mapping <- ID_mapping[!duplicated(ID_mapping$Query),]
  rm(Anno_gene)
  gc()
  genels.remain <- genels[!genels%in%ID_mapping$Query]
  if(!is_empty(genels.remain)){
    PubMedEntrezRes <- GetGeneAnno_Pubmed(
      genels.remain, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    rm(genels.remain)
    gc()
    ID_mapping <- ID_mapping[,-c("UniqueID")]
    ID_mapping <- rbind(
      ID_mapping, PubMedEntrezRes
    )
  }
  ID_mapping <- ID_mapping[,-c("UniqueID")]
  ID_mapping <- ID_mapping[!duplicated(ID_mapping$Query),]
  colnames(ID_mapping)[1] <- "GeneSymbol"
  ID_mapping <- ID_mapping[order(ID_mapping$ensembl_gene_id),]
  ID_mapping <- as.data.frame(ID_mapping)
  ID_mapping <- ID_mapping[!is.na(ID_mapping$GeneSymbol),]
  rownames(ID_mapping)<-ID_mapping$GeneSymbol
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Get Uniprot Annotation")}
  Query <- ID_mapping$GeneSymbol[is.na(ID_mapping$ensembl_gene_id)&is.na(ID_mapping$entrezgene_id)]
  if(!is_empty(Query)){
    UniProtAnno <- GetGeneAnno_UniProt(
      Query = Query, Par_species.Name = Par_species.Name,
      Par_species.Taxid = Par_species.Taxid, dataset = dataset
    )
    ID_mapping <-rbind(
      ID_mapping[!ID_mapping$GeneSymbol%in%UniProtAnno$GeneSymbol,],
      UniProtAnno
    )
    ID_mapping <- ID_mapping[order(ID_mapping$ensembl_gene_id),]
    ID_mapping <- as.data.frame(ID_mapping)
    rownames(ID_mapping)<-ID_mapping$GeneSymbol
    rm(Query, UniProtAnno)
    gc()
  }
  EntrezID <- ID_mapping[!is.na(ID_mapping$entrezgene_id),]
  EntrezID <- EntrezID[order(EntrezID$ensembl_gene_id),]
  EntrezID <- as.data.frame(EntrezID)
  rownames(EntrezID) <- EntrezID$GeneSymbol
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Get Gene Summary")}
  Anno_Summary<-NULL
  attemp<-1
  while ( is.null(Anno_Summary) && attemp<10) {
    try({
      Anno_Summary<-getGenes(geneid = EntrezID$entrezgene_id , fields = c("symbol","summary","generif"))
    })
    attemp<-attemp+1
    if(is.null(Anno_Summary)){Sys.sleep(10)}
  }
  temp<-lapply(
    Anno_Summary$generif,
    FUN = function(x){
      if(is.null(x)){
        return(NA)
      }else{
        substr(paste(apply(x, MARGIN = 1, FUN = paste, collapse = "_"),collapse = " "),0,32700)
      }
    }
  )
  temp<-unlist(temp)
  if(is.null(Anno_Summary@listData$summary)){
    summary<-rep("NA",length(Anno_Summary@listData$quer))
  }else{
    summary<-Anno_Summary@listData$summary
  }
  Anno_Summary<-data.frame(
    entrezgene_id=Anno_Summary@listData$query,
    summary=summary,
    generif=temp
  )
  Anno_Summary<-Anno_Summary[!duplicated(Anno_Summary$entrezgene_id),]
  rownames(Anno_Summary)<-Anno_Summary$entrezgene_id
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Get Gene Ontology")}
  GeneOntology<-data.frame()
  Par_cl<-makeCluster(parallelly::freeConnections()-1)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=1, max=length(rownames(ID_mapping)), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  #Refer here for all column names for Uniport API (https://www.uniprot.org/help/uniprotkb_column_names)
  GeneOntology<-foreach(
    i=1:length(ID_mapping$ensembl_gene_id), .combine=dplyr::bind_rows,
    .options.snow=Par_opts, .packages = "dplyr", .errorhandling = "remove"
  ) %dopar% {
    as.data.frame(
      cbind(
        ensembl_gene_id=ID_mapping$ensembl_gene_id[i],
        data.table::fread(
          paste0(
            "https://rest.uniprot.org/uniprotkb/search?query=",
            ID_mapping$ensembl_gene_id[i],
            "+organism_id:",Par_species.Taxid,
            "&format=tsv&fields=accession,keyword,cc_function,go,xref_kegg,cc_pathway"
          ),verbose=FALSE,showProgress=TRUE
        )[1,]
      )
    )%>%return()
  }
  stopCluster(Par_cl)
  GeneOntology <- GeneOntology[!duplicated(GeneOntology$ensembl_gene_id),]
  GeneOntology <- GeneOntology[!is.na(GeneOntology$ensembl_gene_id),]
  rownames(GeneOntology) <- GeneOntology$ensembl_gene_id
  Par_cl<-makeCluster(parallelly::freeConnections()-3)
  registerDoSNOW(Par_cl)
  Par_pb <- txtProgressBar(min=0, max=nrow(ID_mapping), style = 3)
  progress <- function(n) setTxtProgressBar(Par_pb, n)
  Par_opts<-list(progress = progress)
  ## Refer here for all column names for
  ## Entrez API (https://www.ncbi.nlm.nih.gov/books/NBK25501/)
  ID_mapping<-foreach(
    i=1:nrow(ID_mapping), .errorhandling = "remove",
    .inorder=FALSE, .options.snow=Par_opts,
    .packages = c("utils","S4Vectors", "data.table"),
    .noexport = c("temp")
  ) %dopar% {
    temp <- ID_mapping[i,]
    return(
      data.table(cbind(temp, Anno_Summary[temp$ensembl_gene_id,-1], GeneOntology[temp$ensembl_gene_id,-1]))
    )
  }
  stopCluster(Par_cl)
  ID_mapping <- rbindlist(ID_mapping)
  ID_mapping <- ID_mapping[!duplicated(ID_mapping$GeneSymbol),]
  ID_mapping <- ID_mapping[order(ID_mapping$ensembl_gene_id),]
  ID_mapping <- as.data.frame(ID_mapping)
  rownames(ID_mapping)<-ID_mapping$GeneSymbol
  EntrezID <- ID_mapping[!is.na(ID_mapping$entrezgene_id),]
  EntrezID <- EntrezID[order(EntrezID$ensembl_gene_id),]
  rownames(EntrezID) <- EntrezID$GeneSymbol
  colnames(ID_mapping)[c(7:14)] <- c(
    "RefSeqSummary", "GeneRif", "UniProtKBID", "UniProtKeywords",
    "UniProtFunction", "GO", "KEGG", "UniProtPathway"
  )
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Finishing")}
  pattern<-'(\\\\r)|(\\\\n)|(\\\\a)|(\\\\f)|(\\\\v)|(\\\\b)|(\\\\s)|(\\\\t)|(\\\\)'
  ID_mapping$GeneRif <- gsub(pattern,'', ID_mapping$GeneRif)
  ID_mapping$UniProtFunction <- gsub(pattern,'', ID_mapping$UniProtFunction)
  ID_mapping$UniProtPathway <- gsub(pattern,'', ID_mapping$UniProtPathway)
  ID_mapping$RefSeqSummary <- gsub(pattern,'', ID_mapping$RefSeqSummary)
  ID_mapping$GO <- gsub(pattern,'', ID_mapping$GO)
  ID_mapping$KEGG <- gsub(pattern,'', ID_mapping$KEGG)
  ID_mapping$GeneRif <- substr(gsub('"',"'", ID_mapping$GeneRif),0,32700)
  ID_mapping$gene_biotype <- factor(
    x = as.character(ID_mapping$gene_biotype), levels = unique(ID_mapping$gene_biotype)
  )
  if(ShinyProgress){incProgress(ShinyProg.N/steps, detail = "Done")}
  GeneAnno <- list(ID_mapping=ID_mapping, EntrezID=EntrezID)
  return(GeneAnno)
}
Rtrend <- function(x) {
  if (x > 0) {
    return(1)
  } else if (x < 0) {
    return(-1)
  } else if (x == 0) {
    return(0)
  }
}
pattern <- function(x) {
  for (l in 1:length(x)) {
    x[l] <- Rtrend(x[l])
  }
  return(x)
}
GOMostDescend2 <- function(x) {
  Par_GOCC <- as.list(GOCCOFFSPRING)
  Par_GOBP <- as.list(GOBPOFFSPRING)
  Par_GOMF <- as.list(GOMFOFFSPRING)
  GOlist <- c(Par_GOCC, Par_GOBP, Par_GOMF)
  for (i in length(x):1) {
    if (sum(x %in% GOlist[[x[i]]]) != 0) {
      x <- x[-i]
    }
  }
  return(x)
}
GOMostDescend <- function(x) {
  Par_GOCC <- as.list(GOCCANCESTOR)
  Par_GOBP <- as.list(GOBPANCESTOR)
  Par_GOMF <- as.list(GOMFANCESTOR)
  GOlist <- c(Par_GOCC, Par_GOBP, Par_GOMF)
  for (i in 1:length(x)) {
    x <- x[!(x %in% GOlist[[x[i]]])]
  }
  return(x)
}
GOMostAscend <- function(x, GOlist) {
  while (i <= length(x)) {
    x <- x[!(x %in% GOlist[[x[i]]])]
    i <- i + 1
  }
  return(x)
}
lookup.wzy <- function(a, b, cola) {
  temp <- c()
  for (i in 1:nrow(a)) {
    temp <- rbind(temp, b[rownames(b) == as.character(a[i, cola]), ])
  }
  a <- cbind(a, temp)
  return(a)
}
Df4Plot <- function(
    Gene, Type = "log2TPM", RUV_x = RUV_x,
    DEGs_Out_TPM = DEGs_Out_TPM, colname = "external_gene_name"
    ) {
  temp <- as.numeric(DEGs_Out_TPM[
    DEGs_Out_TPM[, colname] == Gene,
    c(grep(Type, colnames(DEGs_Out_TPM), value = TRUE))
  ])
  temp <- data.frame(as.character(RUV_x), as.numeric(temp))
  colnames(temp) <- c("x", Type)
  temp <- as.data.frame(temp)
  temp$x <- factor(temp$x, unique(temp$x))
  return(temp)
}
BinMean <- function(vec, every, na.rm = FALSE) {
  n <- length(vec)
  x <- .colMeans(vec, every, n %/% every, na.rm)
  r <- n %% every
  if (r) x <- c(x, mean.default(vec[(n - r + 1):n], na.rm = na.rm))
  x
}
BinMean2 <- function(vec, every, na.rm = FALSE) {
  start <- c(1, head(cumsum(every) + 1, -1))
  end <- cumsum(every)
  x <- c()
  for (i in 1:length(every)) {
    x <- c(x, mean(vec[start[i]:end[i]]))
  }
  x
}
FUN <- function(x) {
  if (is.null(x)) {
    return(NA)
  } else {
    substr(paste(apply(x, MARGIN = 1, FUN = paste, collapse = "_"), collapse = " "), 0, 32700)
  }
}
wzy.2power <- function(x) {
  if (is.null(dim(x))) {
    x[x == 0] <- -Inf
  } else {
    x <- apply(x, 2, function(x) {
      x[x == 0] <- -Inf
      return(x)
    })
  }
  x <- 2^x
  return(x)
}
ping <- function(x, stderr = FALSE, stdout = FALSE, ...) {
  pingvec <- system2("ping", x,
                     stderr = FALSE,
                     stdout = FALSE, ...
  )
  if (pingvec == 0) TRUE else FALSE
}
MeanRetrieve <- function(ID, Anno2, DEGs_log2TPM, RUV_x, Par_Rep) {
  if (sum(Anno2$ensembl_gene_id %in% ID)) {
    ID <- rownames(Anno2)[Anno2$ensembl_gene_id %in% ID]
  } else {
    ID <- rownames(Anno2)[Anno2$entrezgene_id %in% ID]
  }
  if (sum(rownames(DEGs_log2TPM) %in% ID) == 1) {
    temp <- DEGs_log2TPM %>%
      apply(., 2, wzy.2power) %>%
      .[rownames(.) %in% ID, ] %>%
      c(BinMean2(., Par_Rep), .) %>%
      log(., 2)
  } else {
    temp <- DEGs_log2TPM %>%
      apply(., 2, wzy.2power) %>%
      .[rownames(.) %in% ID, ] %>%
      colMeans(.) %>%
      c(BinMean2(., Par_Rep), .) %>%
      log(., 2)
  }
  names(temp) <- ifelse(names(temp) == "", paste0(levels(RUV_x), "_Log2MeanTPM"), names(temp))
  return(temp)
}
GSEA_IDconvert <- function(tb4plot, ID_mapping, WD, lv1, ChkSlt_Lv2, ChkSlt_Lv4) {
  test0 <- gsubfn::strapplyc(as.character(tb4plot$core_enrichment), pattern = "^(ENS[[:alpha:]]*).*")
  test <- gsubfn::strapplyc(as.character(tb4plot$core_enrichment), pattern = paste0("^", test0[[1]], "([-0-9]+)"))
  if (sum(isEmpty(test0))) {
    keyType <- "ENTREZID"
  } else {
    keyType <- "ENSEMBL"
  }
  if (keyType != "ENSEMBL") {
    for (rn in 1:nrow(tb4plot)) {
      entrez_id2 <- ID_mapping$external_gene_name
      names(entrez_id2) <- ID_mapping$entrezgene_id
      rn_tmp <- tb4plot$core_enrichment[rn] %>%
        str_split(., "/") %>%
        .[[1]] %>%
        entrez_id2[.]
      rn_tmp[rn_tmp %in% ""] <- tb4plot$core_enrichment[rn] %>%
        str_split(., "/") %>%
        unlist() %>%
        .[rn_tmp %in% ""]
      tb4plot$core_enrichment[rn] <- rn_tmp %>% paste(., collapse = "|")
    }
  } else {
    for (rn in 1:nrow(tb4plot)) {
      if (ChkSlt_Lv4 == "KEGG") {
        entrez_id2 <- ID_mapping$external_gene_name
        names(entrez_id2) <- ID_mapping$entrezgene_id
        rn_tmp <- tb4plot$core_enrichment[rn] %>%
          str_split(., "/") %>%
          .[[1]] %>%
          entrez_id2[.]
        rn_tmp[rn_tmp %in% ""] <- tb4plot$core_enrichment[rn] %>%
          str_split(., "/") %>%
          unlist() %>%
          .[rn_tmp %in% ""]
        tb4plot$core_enrichment[rn] <- rn_tmp %>% paste(., collapse = "|")
      } else {
        ensembl_id <- ID_mapping$external_gene_name
        names(ensembl_id) <- ID_mapping$ensembl_gene_id
        rn_tmp <- tb4plot$core_enrichment[rn] %>%
          str_split(., "/") %>%
          .[[1]] %>%
          ensembl_id[.]
        rn_tmp[rn_tmp %in% ""] <- tb4plot$core_enrichment[rn] %>%
          str_split(., "/") %>%
          unlist() %>%
          .[rn_tmp %in% ""]
        tb4plot$core_enrichment[rn] <- rn_tmp %>% paste(., collapse = "|")
      }
    }
  }
  ID_list<-list.files(file.path(WD, "01.PlotOutput", lv1, ChkSlt_Lv2, ChkSlt_Lv4))%>%
    grep("^GSEA_.*\\.jpeg$",.,value=TRUE)
  ID_list_ID<-ID_list%>%str_extract(.,"_[:alpha:]*\\.{0,1}[0-9]*[^_]")%>%str_remove(.,"_")%>%str_replace(.,"\\.",":")
  ID_list<-ID_list[ID_list_ID%in%rownames(tb4plot)]
  ID_list_ID<-ID_list_ID[ID_list_ID%in%rownames(tb4plot)]
  tb4plot[ID_list_ID, "leading_edge"]<-hmakeTag(
    'a', tb4plot[ID_list_ID,"leading_edge"],
    href=paste0(file.path("01.PlotOutput", lv1, ChkSlt_Lv2, ChkSlt_Lv4),.Platform$file.sep,ID_list),target='_blank'
  )
  if(ChkSlt_Lv4=="KEGG"){
    ID_list<-list.files(file.path(WD, "01.PlotOutput", lv1, ChkSlt_Lv2, ChkSlt_Lv4)) %>%
      grep("\\.pathview\\.png$",.,value=TRUE)
    ID_list<-ID_list[(ID_list%>%str_extract(.,"^[:alpha:]*[0-9]*[^\\.]"))%in%tb4plot$ID]
    tb4plot[ID_list%>%str_extract(.,"^[:alpha:]*[0-9]*[^\\.]"),"ID"]<-hmakeTag(
      'a', tb4plot[ID_list%>%str_extract(.,"^[:alpha:]*[0-9]*[^\\.]"),"ID"],
      href=paste0(file.path("01.PlotOutput",lv1, ChkSlt_Lv2, ChkSlt_Lv4), .Platform$file.sep, ID_list),target='_blank'
    )
  }
  return(tb4plot)
}
ORA_IDconvert <- function(tb4plot, ID_mapping, WD, lv1, ChkSlt_Lv2, lv3, ChkSlt_Lv4) {
  test0 <- gsubfn::strapplyc(as.character(tb4plot$geneID), pattern = "^(ENS[[:alpha:]]*).*")
  test <- gsubfn::strapplyc(as.character(tb4plot$geneID), pattern = paste0("^", test0[[1]], "([-0-9]+)"))
  if (sum(isEmpty(test0))) {
    keyType <- "ENTREZID"
  } else {
    keyType <- "ENSEMBL"
  }
  if (keyType != "ENSEMBL") {
    for (rn in 1:nrow(tb4plot)) {
      entrez_id2 <- ID_mapping$external_gene_name
      names(entrez_id2) <- ID_mapping$entrezgene_id
      rn_tmp <- tb4plot$geneID[rn] %>%
        str_split(., "/") %>%
        .[[1]] %>%
        entrez_id2[.]
      rn_tmp[rn_tmp %in% ""] <- tb4plot$geneID[rn] %>%
        str_split(., "/") %>%
        unlist() %>%
        .[rn_tmp %in% ""]
      tb4plot$geneID[rn] <- rn_tmp %>% paste(., collapse = "|")
    }
  } else {
    for (rn in 1:nrow(tb4plot)) {
      if (ChkSlt_Lv4 == "KEGG") {
        entrez_id2 <- ID_mapping$external_gene_name
        names(entrez_id2) <- ID_mapping$entrezgene_id
        rn_tmp <- tb4plot$geneID[rn] %>%
          str_split(., "/") %>%
          .[[1]] %>%
          entrez_id2[.]
        rn_tmp[rn_tmp %in% ""] <- tb4plot$geneID[rn] %>%
          str_split(., "/") %>%
          unlist() %>%
          .[rn_tmp %in% ""]
        tb4plot$geneID[rn] <- rn_tmp %>% paste(., collapse = "|")
      } else {
        ensembl_id <- ID_mapping$external_gene_name
        names(ensembl_id) <- ID_mapping$ensembl_gene_id
        rn_tmp <- tb4plot$geneID[rn] %>%
          str_split(., "/") %>%
          .[[1]] %>%
          ensembl_id[.]
        rn_tmp[rn_tmp %in% ""] <- tb4plot$geneID[rn] %>%
          str_split(., "/") %>%
          unlist() %>%
          .[rn_tmp %in% ""]
        tb4plot$geneID[rn] <- rn_tmp %>% paste(., collapse = "|")
      }
    }
  }
  if (ChkSlt_Lv4 == "KEGG") {
    ID_list <- list.files(file.path(WD, "01.PlotOutput", lv1, ChkSlt_Lv2, lv3)) %>%
      grep("\\.pathview\\.png$", ., value = TRUE)
    ID_list <- ID_list[(ID_list %>% str_extract(., "^[:alpha:]*[0-9]*[^\\.]")) %in% tb4plot$ID]
    tb4plot[ID_list %>% str_extract(., "^[:alpha:]*[0-9]*[^\\.]"), "ID"] <- hmakeTag(
      "a", tb4plot[ID_list %>% str_extract(., "^[:alpha:]*[0-9]*[^\\.]"), "ID"],
      href = paste0(file.path("01.PlotOutput", lv1, ChkSlt_Lv2, lv3), .Platform$file.sep, ID_list), target = "_blank"
    )
  }
  return(tb4plot)
}
WZY_RunFEA <- function(
    log2FColNa = "avg_log2FC", ColChk = TRUE, Group.By = "cluster", Data = Data,
    tbOut = Data, WD = WD, FilePath = "Data_mg_FEA_CMarkers", ShinyProgress = FALSE,
    AnnoHub = "org.Mm.eg.db", Table1_rows_all = rownames(Data)[Data$power>0.4],
    Par_UseFDR4FEA = FALSE, Par_FDR4FEA = 0.05, Par_FDR4KEG_OR = 0.05,
    Par_species.go.kegg = "mmu", descend = TRUE
){
  library(gtools)
  library(BiocParallel)
  library(clusterProfiler)
  library(pathview)
  library(enrichplot)
  library(GO.db)
  library(DOSE)
  library(tidytree)
  library(treeio)
  library(ggtree)
  
  Folder <- FilePath %>% dirname()
  FileName <- FilePath %>% basename() %>% str_remove(.,"\\.xlsx$") %>% substr(., 1, 24)
  WD <- file.path(Folder, paste0(FileName, "_FEA_Res"))
  
  if(!dir.exists(WD)){
    dir.create(WD)
  }
  if(file.exists(file.path(WD, "01.PlotOutput"))){
    unlink(file.path(WD, "01.PlotOutput"), recursive = TRUE)
    dir.create(file.path(WD, "01.PlotOutput"))
  }else{
    dir.create(file.path(WD, "01.PlotOutput"))
  }
  if(file.exists(file.path(WD, "02.TableOutput"))){
    unlink(file.path(WD, "02.TableOutput"), recursive = TRUE)
    dir.create(file.path(WD, "02.TableOutput"))
  }else{
    dir.create(file.path(WD, "02.TableOutput"))
  }
  if(!AnnoHub %in% installed.packages()){
    BiocManager::install(AnnoHub, update = FALSE)
  }
  invisible(lapply(AnnoHub, library, character.only = TRUE))
  assign("edb",get(AnnoHub))
  RunningData <- tbOut[Table1_rows_all,]
  if(Group.By!="NULL"){
    RunningData[ ,Group.By]<-factor(
      RunningData[ ,Group.By],
      levels = unique(RunningData[ ,Group.By])
    )
    Par_ConLs_C<-levels(RunningData[ ,Group.By])%>% mixedsort()
  }else{
    Par_ConLs_C<-basename(WD)
  }
  DEGs_ls<-list()
  if(ShinyProgress){iincProgress(0.1, detail = "Running FEA for each group")}
  for (i in 1:length(Par_ConLs_C)){
    #|----Construction----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Construction")}else{
      print(paste0(Sys.time(), ": Processing Cluster ", Par_ConLs_C[i]))
    }
    Sheet1<-Par_ConLs_C[i] %>% substr(., 1, 8)
    
    if(Group.By!="NULL"){
      SltRows_All<-tbOut[,Group.By]==Par_ConLs_C[i]
      SltRows_Thres<-RunningData[,Group.By]==Par_ConLs_C[i]
    }else{
      SltRows_All<-rep(TRUE, nrow(RunningData))
      SltRows_Thres<-SltRows_All
    }
    
    tb_is_sig_ensembl<-RunningData[SltRows_Thres,]$ensembl_gene_id
    
    DEGs_Div<-list(
      DEGs_raw=tbOut[SltRows_All,],
      GSEA_Enrich=list(GO_BP=list(),GO_MF=list(),GO_CC=list(),KEGG=list()),
      Over_Represent=list(
        All=list(GO_BP=list(),GO_MF=list(),GO_CC=list(),KEGG=list()),
        Up_regulated=list(GO_BP=list(),GO_MF=list(),GO_CC=list(),KEGG=list()),
        Down_regulated=list(GO_BP=list(),GO_MF=list(),GO_CC=list(),KEGG=list())
      )
    )
    DEGs_Div$DEGs_raw<-cbind(
      DEGs_Div$DEGs_raw,
      CutOff=rep(0,nrow(DEGs_Div$DEGs_raw))
    )
    DEGs_Div$DEGs_raw$CutOff[
      DEGs_Div$DEGs_raw$ensembl_gene_id%in%tb_is_sig_ensembl
    ]<-1
    DEGs_res_out<-DEGs_Div$DEGs_raw
    #|----Functional Enrichment----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Start FEA")}
    #|----Prepare folder----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Prepare folder")}
    FolderName<-paste0("FEA_Plot_",sprintf("%02d", i),"_",Par_ConLs_C[i])
    dir.create(file.path(WD,"01.PlotOutput",FolderName), showWarnings = FALSE)
    if(file.exists(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich"))){
      unlink(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich"), recursive=TRUE)
    }
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","GO_BP"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","GO_MF"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","GO_CC"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","KEGG"),
               showWarnings = FALSE)
    if(file.exists(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent"))){
      unlink(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent"),
             recursive=TRUE)
    }
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",
                         "All_KEGG"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",
                         "Up_regulated_KEGG"),
               showWarnings = FALSE)
    dir.create(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",
                         "Down_regulated_KEGG"),
               showWarnings = FALSE)
    #|----Prepare data for GSEA enrichment----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Prepare Data for GSEA")}
    FEA_GeneLs <- DEGs_Div$DEGs_raw[!is.na(DEGs_Div$DEGs_raw$entrezgene_id),]
    FEA_GeneLsFC <- FEA_GeneLs[, log2FColNa]
    names(FEA_GeneLsFC) <- FEA_GeneLs$entrezgene_id
    FEA_GeneLsFC[FEA_GeneLsFC==Inf]<-max(
      FEA_GeneLsFC[is.finite(FEA_GeneLsFC)]
    )%>%ceiling()
    FEA_GeneLsFC[FEA_GeneLsFC==-Inf]<-min(
      FEA_GeneLsFC[is.finite(FEA_GeneLsFC)]
    )%>%floor()
    FEA_GeneLsFC_EMBL <- DEGs_Div$DEGs_raw[
      !is.na(DEGs_Div$DEGs_raw$ensembl_gene_id),
    ]
    temp <- FEA_GeneLsFC_EMBL[, log2FColNa]
    names(temp) <- FEA_GeneLsFC_EMBL$ensembl_gene_id
    FEA_GeneLsFC_EMBL <- temp
    FEA_GeneLsFC_EMBL<-sort(FEA_GeneLsFC_EMBL, decreasing = TRUE)
    FEA_GeneLsFC_EMBL[FEA_GeneLsFC_EMBL==Inf]<-max(
      FEA_GeneLsFC_EMBL[is.finite(FEA_GeneLsFC_EMBL)]
    )%>%ceiling()
    FEA_GeneLsFC_EMBL[FEA_GeneLsFC_EMBL==-Inf]<-min(
      FEA_GeneLsFC_EMBL[is.finite(FEA_GeneLsFC_EMBL)]
    )%>%floor()
    rm(temp)
    gc()
    #|----get log2 mean fold change for duplicated IDs
    Dup_IDs<-unique(names(FEA_GeneLsFC)[duplicated(names(FEA_GeneLsFC))])
    if(is_empty(Dup_IDs)){
      FEA_GeneLsFC<-sort(FEA_GeneLsFC, decreasing = TRUE)
    }else{
      temp1<-FEA_GeneLsFC[!(names(FEA_GeneLsFC)%in%Dup_IDs)]
      temp2<-FEA_GeneLsFC[names(FEA_GeneLsFC)%in%Dup_IDs]
      for (m in 1:length(Dup_IDs)) {
        temp3<-log(mean(wzy.2power(temp2[names(temp2)==Dup_IDs[m]])),2)
        names(temp2)<-Dup_IDs[m]
        temp1<-c(
          temp1,
          temp3
        )
      }
      FEA_GeneLsFC<-sort(temp1, decreasing = TRUE)
    }
    Dup_IDs<-unique(names(FEA_GeneLsFC_EMBL)[duplicated(names(FEA_GeneLsFC_EMBL))])
    if(is_empty(Dup_IDs)){
      FEA_GeneLsFC_EMBL<-sort(FEA_GeneLsFC_EMBL, decreasing = TRUE)
    }else{
      temp1<-FEA_GeneLsFC_EMBL[!(names(FEA_GeneLsFC_EMBL)%in%Dup_IDs)]
      temp2<-FEA_GeneLsFC_EMBL[names(FEA_GeneLsFC_EMBL)%in%Dup_IDs]
      for (m in 1:length(Dup_IDs)) {
        temp3<-log(mean(wzy.2power(temp2[names(temp2)==Dup_IDs[m]])),2)
        names(temp2)<-Dup_IDs[m]
        temp1<-c(
          temp1,
          temp3
        )
      }
      FEA_GeneLsFC_EMBL<-sort(temp1, decreasing = TRUE)
    }
    #|----Gene Ontology (GO) enrichment (GSEA enrichment)----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "GSEA:GO")}
    if(sum(keytypes(edb)%in%"ENSEMBL")){
      keyType<-"ENSEMBL"
    }else{
      keyType<-"ENTREZID"
      FEA_GeneLsFC_EMBL<-FEA_GeneLsFC
    }
    
    for (j in c("BP","MF","CC")) {
      FEA_GO <- gseGO(geneList     = FEA_GeneLsFC_EMBL,
                      OrgDb        = edb,
                      keyType      = keyType,
                      nPermSimple  = 30000,
                      eps          = 1e-10,
                      pAdjustMethod = "BH",
                      ont          = j,
                      minGSSize    = 10,
                      maxGSSize    = 2000,
                      pvalueCutoff = 1,
                      seed         = FALSE,
                      BPPARAM=MulticoreParam(workers = 1),
                      verbose      = TRUE)
      if(is.null(FEA_GO)){next()}
      if(Par_UseFDR4FEA==FALSE){
        FEA_GO@result<-FEA_GO@result[FEA_GO@result$pvalue<Par_FDR4FEA,]
      }else{
        FEA_GO@result<-FEA_GO@result[FEA_GO@result$p.adjust<Par_FDR4FEA,]
      }
      if(nrow(FEA_GO)>0){
        #|----Only The most descend term----|####
        if(descend){
          FEA_GO_MostDescend<-GOMostDescend(x=FEA_GO@result$ID)
          FEA_GO_MostDescend<-GOMostDescend2(x=FEA_GO_MostDescend)
          FEA_GO@result<-FEA_GO@result[FEA_GO@result$ID%in%FEA_GO_MostDescend,]
          FEA_GO@geneSets<-FEA_GO@geneSets[
            names(FEA_GO@geneSets)%in%FEA_GO_MostDescend
          ]
        }
        #|----Pass results to collection----|####
        DEGs_Div$GSEA_Enrich[[match(j,c("BP","MF","CC"))]]<-list(
          Raw=FEA_GO,Sig=FEA_GO@result
        )
        names(DEGs_Div$GSEA_Enrich)[match(j,c("BP","MF","CC"))]<-paste0("GO_",j)
        if(nrow(FEA_GO)>50){
          FEA_title<-paste0(
            "Top 50 significant GO_", j,
            " terms (", Par_ConLs_C[i], "; the most descend terms)"
          )
        }else if(nrow(FEA_GO)<=50){
          FEA_title<-paste0(
            "All ", nrow(FEA_GO)," significant GO_", j,
            " terms (", Par_ConLs_C[i], "; the most descend terms)"
          )
        }
        #|----Plot Out----|####
        if(nrow(FEA_GO)>10){
          p1<-dotplot(FEA_GO, showCategory=50, label_format=100, color = "pvalue")
          p2<-treeplot(pairwise_termsim(FEA_GO), showCategory=50, 
                       cluster.params = list(method = "ward.D2"), color = "pvalue")
          svg(
            file.path(
              WD,"01.PlotOutput",FolderName,"GSEA_Enrich",paste0("gseGO_",j,".svg")
            ), 
            width = 24, height = 10, pointsize = 12
          )
          gridExtra::grid.arrange(
            p1, p2, nrow = 1, widths = c(1, 1),
            top=textGrob(paste0("GSEA: ",FEA_title," with similarity summary"))
          )
          dev.off()
          graphics.off()
          rm(p1,p2)
        }else{
          p1<-dotplot(FEA_GO, showCategory=50, label_format=100, color = "pvalue")+
            ggtitle(paste0("GSEA: ",FEA_title))
          svg(
            file.path(
              WD,"01.PlotOutput",FolderName,"GSEA_Enrich",paste0("gseGO_",j,".svg")
            ), 
            width = 24, height = 10, pointsize = 12
          )
          print(p1)
          dev.off()
          graphics.off()
          rm(p1)
        }
        FEA_GO@geneSets<-FEA_GO@geneSets[
          (FEA_GO@geneSets%>%names())%in%FEA_GO@result$ID
        ]
        for (k in 1:nrow(FEA_GO)) {
          jpeg(
            file.path(
              WD,"01.PlotOutput", 
              FolderName,"GSEA_Enrich",
              paste0("GO_",j),
              paste0(
                substr(
                  make.names(
                    paste0("GSEA_",paste0(FEA_GO@result[k,1:2],collapse = "_"))
                  ),0,65
                ),".jpeg"
              )
            ),  
            width = 900, height = 600, pointsize = 14, type = "cairo"
          )
          gplot<-gseaplot2(
            FEA_GO, pvalue_table=FALSE,
            geneSetID=which(names(FEA_GO@geneSets)%in%FEA_GO@result$ID[k]),
            title = FEA_GO@result$Description[k]
          )
          print(gplot)
          dev.off()
          graphics.off()
        }
      }else{
        DEGs_Div$GSEA_Enrich[[match(j,c("BP","MF","CC"))]]<-list(
          Raw=list(),Sig=list()
        )
        names(DEGs_Div$GSEA_Enrich)[match(j,c("BP","MF","CC"))]<-paste0("GO_",j)
      }
    }
    #|----KEGG pathway enrichment (GSEA enrichment)----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "GSEA:KEGG")}
    FEA_KEGG <- gseKEGG(geneList     = FEA_GeneLsFC,
                        organism     = Par_species.go.kegg,
                        nPermSimple  = 30000,
                        eps          = 1e-10,
                        minGSSize    = 80,
                        pvalueCutoff = 1,
                        BPPARAM=MulticoreParam(workers = 1),
                        verbose      = TRUE)
    
    if(is.null(FEA_KEGG)){
      DEGs_Div$GSEA_Enrich$KEGG<-list(Raw=list(),Sig=list())
    }else{
      if(Par_UseFDR4FEA==FALSE){
        FEA_KEGG@result<-FEA_KEGG@result[FEA_KEGG@result$pvalue<Par_FDR4FEA,]
      }else{
        FEA_KEGG@result<-FEA_KEGG@result[FEA_KEGG@result$p.adjust<Par_FDR4FEA,]
      }
      if(nrow(FEA_KEGG)>0){
        FEA_KEGG@geneSets<-FEA_KEGG@geneSets[
          names(FEA_KEGG@geneSets)%in%FEA_KEGG@result$ID
        ]
        #|----Pass results to collection----|####
        DEGs_Div$GSEA_Enrich$KEGG<-list(Raw=FEA_KEGG, Sig=FEA_KEGG@result)
        if(nrow(FEA_KEGG)>50){
          FEA_title<-paste0(
            "Top 50 significant KEGG pathways (", Par_ConLs_C[i], ")"
          )
        }else if(nrow(FEA_KEGG)<=50){
          FEA_title<-paste0(
            "All ", nrow(FEA_KEGG)," significant KEGG pathways (",
            Par_ConLs_C[i], ")"
          )
        }
        #|----Plot Out----|####
        if(nrow(FEA_KEGG)>10){
          p1<-dotplot(FEA_KEGG, showCategory=50, label_format=100, color = "pvalue")
          p2<-treeplot(
            pairwise_termsim(FEA_KEGG), showCategory=50, nWords = 0,
            cluster.params = list(method = "ward.D2"), color = "pvalue"
          )
          svg(
            file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","gseKEGG.svg"), 
            width = 24, height = 10, pointsize = 12
          )
          gridExtra::grid.arrange(p1, p2, nrow = 1, widths = c(1, 1),
                                  top=textGrob(paste0("GSEA: ",FEA_title," with similarity tree plot")))
          dev.off()
          rm(p1,p2)
        }else{
          p1<-dotplot(FEA_KEGG, showCategory=50, label_format=100, color = "pvalue")+
            ggtitle(paste0("GSEA: ",FEA_title))
          svg(
            file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","gseKEGG.svg"), 
            width = 24, height = 10, pointsize = 12
          )
          print(p1)
          dev.off()
          rm(p1)
        }
        for (k in 1:nrow(FEA_KEGG@result)) {
          jpeg(
            filename = file.path(
              WD,"01.PlotOutput", 
              FolderName,"GSEA_Enrich","KEGG",
              paste0(substr(make.names(paste0("GSEA_",paste0(FEA_KEGG@result[k,1:2],collapse = "_"))),0,65),".jpeg")
            ),  
            width = 900, height = 600, pointsize = 14, type = "cairo"
          )
          gplot<-gseaplot2(FEA_KEGG, pvalue_table=FALSE,
                           geneSetID=which(names(FEA_KEGG@geneSets)%in%FEA_KEGG@result$ID[k]),
                           title = FEA_KEGG@result$Description[k])
          print(gplot)
          dev.off()
          wd<-getwd()
          setwd(file.path(WD,"01.PlotOutput",FolderName,"GSEA_Enrich","KEGG"))
          
          tryCatch({
            pathview(gene.data=FEA_GeneLsFC[names(FEA_GeneLsFC)%in%(FEA_KEGG@result$core_enrichment[k]%>%str_split(.,"/")%>%.[[1]])],
                     pathway.id=FEA_KEGG@result$ID[k],
                     species=Par_species.go.kegg)
          }, error=function(e){}, finally = c())
          setwd(wd)
        }
      }else{
        DEGs_Div$GSEA_Enrich$KEGG<-list(Raw=list(),Sig=list())
      }
    }
    #|----Over Representation Test----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "ORA:GO")}
    for (l in c("All","Up_regulated","Down_regulated")) {
      if(l=="All"){
        gene<-FEA_GeneLs$entrezgene_id[FEA_GeneLs$CutOff==1]
        gene<-unique(gene)
        gene_EMBL<-DEGs_res_out$ensembl_gene_id[DEGs_res_out$CutOff==1]
        gene_EMBL<-unique(gene_EMBL)
      }else if(l=="Up_regulated"){
        gene<-FEA_GeneLs$entrezgene_id[FEA_GeneLs$CutOff==1 & FEA_GeneLs[,log2FColNa]>0]
        gene<-unique(gene)
        gene_EMBL<-DEGs_res_out$ensembl_gene_id[DEGs_res_out$CutOff==1 & DEGs_res_out[,log2FColNa]>0]
        gene_EMBL<-unique(gene_EMBL)
      }else if(l=="Down_regulated"){
        gene<-FEA_GeneLs$entrezgene_id[FEA_GeneLs$CutOff==1 & FEA_GeneLs[,log2FColNa]<0]
        gene<-unique(gene)
        gene_EMBL<-DEGs_res_out$ensembl_gene_id[DEGs_res_out$CutOff==1 & DEGs_res_out[,log2FColNa]<0]
        gene_EMBL<-unique(gene_EMBL)
      }
      if(keyType!="ENSEMBL"){
        gene_EMBL<-gene
      }
      if(is_empty(gene)){
        for (j in c("BP","MF","CC")) {
          DEGs_Div$Over_Represent[[l]][[match(j,c("BP","MF","CC"))]]<-list(
            Raw=list(),Sig=list()
          )
          DEGs_Div$Over_Represent[[l]]$KEGG<-list(Raw=list(),Sig=list())
        }
      }else{
        #|----Gene Ontology (GO) enrichment (Over Representation Test)----|####
        for (j in c("BP","MF","CC")) {
          FEA_GO_OR <-enrichGO(gene          = gene_EMBL,
                               OrgDb         = edb,
                               keyType       = keyType,
                               ont           = j,
                               pAdjustMethod = "BH",
                               pvalueCutoff  = 1,
                               qvalueCutoff  = 1,
                               readable      = FALSE)
          if(is.null(FEA_GO_OR)){next()}
          if(Par_UseFDR4FEA==FALSE){
            FEA_GO_OR@result<-FEA_GO_OR@result[
              FEA_GO_OR@result$pvalue<Par_FDR4FEA,
            ]
          }else{
            FEA_GO_OR@result<-FEA_GO_OR@result[
              FEA_GO_OR@result$p.adjust<Par_FDR4FEA,
            ]
          }
          if(nrow(FEA_GO_OR)>0){
            #|----Only the most descend term----|####
            if(descend){
              FEA_GO_MostDescend<-GOMostDescend(x=FEA_GO_OR@result$ID)
              FEA_GO_MostDescend<-GOMostDescend2(x=FEA_GO_MostDescend)
              FEA_GO_OR@result<-FEA_GO_OR@result[
                FEA_GO_OR@result$ID%in%FEA_GO_MostDescend,
              ]
              FEA_GO_OR@geneSets<-FEA_GO_OR@geneSets[
                names(FEA_GO_OR@geneSets)%in%FEA_GO_MostDescend
              ]
            }
            #|----Pass results to collection----|####
            DEGs_Div$Over_Represent[[l]][[match(j,c("BP","MF","CC"))]]<-list(
              Raw=FEA_GO_OR,Sig=FEA_GO_OR@result
            )
            if(nrow(FEA_GO_OR)>50){
              FEA_title<-paste0(
                "Top 50 significant GO_", j, " terms (", Par_ConLs_C[i], ")"
              )
            }else if(nrow(FEA_GO_OR)<=50){
              FEA_title<-paste0(
                "All ", nrow(FEA_GO_OR)," significant GO_", j,
                " terms (", Par_ConLs_C[i], ")"
              )
            }
            #|----Plot Out----|####
            if(nrow(FEA_GO_OR)>=25){
              p1<-dotplot(
                FEA_GO_OR, showCategory=50, label_format=100, color = "pvalue"
              )
              p2<-treeplot(
                pairwise_termsim(FEA_GO_OR), showCategory=50,
                cluster.params = list(method = "ward.D2"), color = "pvalue"
              )
              svg(
                file.path(
                  WD,"01.PlotOutput",FolderName,"Over_Represent",
                  paste0(l, "_GO_",j,".svg")
                ), 
                width = 24, height = 10, pointsize = 12
              )
              gridExtra::grid.arrange(
                p1, p2, nrow = 1, widths = c(1, 1),
                top=textGrob(paste0("Over Representation Test: ",FEA_title,
                                    " with similarity summary from ", l, " DEGs."))
              )
              dev.off()
              rm(p1,p2)
            }else{
              p1<-dotplot(
                FEA_GO_OR, showCategory=50, label_format=100, color = "pvalue"
              )+ggtitle(paste0(
                "Over Representation Test: ",FEA_title," from ", l, " DEGs."
              ))
              svg(
                file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",
                          paste0(l, "_GO_",j,".svg")), 
                width = 24, height = 10, pointsize = 12
              )
              print(p1)
              dev.off()
              rm(p1)
            }
          }else{
            DEGs_Div$Over_Represent[[l]][[match(j,c("BP","MF","CC"))]]<-list(
              Raw=list(),Sig=list()
            )
          }
        }
        #|----KEGG pathway enrichment (Over Representation Test)----|####
        if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "ORA:KEGG")}
        FEA_KEGG_OR <- enrichKEGG(gene      = gene,
                                  organism  = Par_species.go.kegg,
                                  pvalueCutoff = Par_FDR4KEG_OR)
        
        if(is.null(FEA_KEGG_OR)){
          DEGs_Div$Over_Represent[[l]]$KEGG<-list(Raw=list(),Sig=list())
        }else{
          if(Par_UseFDR4FEA==FALSE){
            FEA_KEGG_OR@result<-FEA_KEGG_OR@result[
              FEA_KEGG_OR@result$pvalue<Par_FDR4FEA,
            ]
          }else{
            FEA_KEGG_OR@result<-FEA_KEGG_OR@result[
              FEA_KEGG_OR@result$p.adjust<Par_FDR4FEA,
            ]
          }
          if(nrow(FEA_KEGG_OR)>0){
            FEA_KEGG_OR@geneSets<-FEA_KEGG_OR@geneSets[
              names(FEA_KEGG_OR@geneSets)%in%FEA_KEGG_OR@result$ID
            ]
            #|----Pass results to collection----|####
            DEGs_Div$Over_Represent[[l]]$KEGG<-list(
              Raw=FEA_KEGG_OR,Sig=FEA_KEGG_OR@result
            )
            if(nrow(FEA_KEGG_OR)>50){
              FEA_title<-paste0(
                "Top 50 significant KEGG pathways (", Par_ConLs_C[i], ")"
              )
            }else if(nrow(FEA_KEGG_OR)<=50){
              FEA_title<-paste0(
                "All ", nrow(FEA_KEGG_OR)," significant KEGG pathways (",
                Par_ConLs_C[i], ")"
              )
            }
            #|----Plot Out----|####
            if(nrow(FEA_KEGG_OR)>10){
              p1<-dotplot(
                FEA_KEGG_OR, showCategory=50, label_format=100, color = "pvalue"
              )
              p2<-treeplot(
                pairwise_termsim(FEA_KEGG_OR), showCategory=50, nWords = 0,
                cluster.params = list(method = "ward.D2"), color = "pvalue"
              )
              svg(
                file.path(
                  WD,"01.PlotOutput",FolderName,
                  "Over_Represent",paste0(l,"_KEGG.svg")
                ), 
                width = 24, height = 10, pointsize = 12
              )
              gridExtra::grid.arrange(
                p1, p2, nrow = 1, widths = c(1, 1),
                top=textGrob(paste0(
                  "Over Representation Test: ",FEA_title,
                  " with similarity tree plot from ", l, " DEGs."
                ))
              )
              dev.off()
              rm(p1,p2)
            }else{
              p1<-dotplot(
                FEA_KEGG_OR, showCategory=50, label_format=100, color = "pvalue"
              )+ggtitle(
                paste0("Over Representation Test: ",FEA_title,
                       " from ", l, " DEGs.")
              )
              svg(
                file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",paste0(l,"_KEGG.svg")), 
                width = 24, height = 10, pointsize = 12
              )
              print(p1)
              dev.off()
              rm(p1)
            }
            for (k in 1:nrow(FEA_KEGG_OR@result)) {
              wd<-getwd()
              setwd(file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",paste0(l,"_KEGG")))
              tryCatch({
                pathview(gene.data=FEA_GeneLsFC[names(FEA_GeneLsFC)%in%DEGs_res_out$entrezgene_id[DEGs_res_out$CutOff==1]],
                         pathway.id=FEA_KEGG_OR@result$ID[k],#[!(temp[43:331]%in%c("mmu05206","mmu00511"))], 
                         species=Par_species.go.kegg)
              }, error=function(e){}, finally = c())
              setwd(wd)
            }
          }else{
            DEGs_Div$Over_Represent[[l]]$KEGG<-list(Raw=list(),Sig=list())
          }
        }
      }
    }
    #|----Make Excel .xlsx format output----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Make Excel output")}
    DEGs_Div[["DEGs_raw"]]$CutOff<-ifelse(DEGs_Div[["DEGs_raw"]]$CutOff,"Yes","No")
    colnames(DEGs_Div[["DEGs_raw"]])[ncol(DEGs_Div[["DEGs_raw"]])]<-"Is.Sig."
    DEGs_res_out <- DEGs_Div[["DEGs_raw"]]
    DEGs_res_out <- DEGs_res_out[,!duplicated(colnames(DEGs_res_out))]
    #|----All DEGs output----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "All DEGs output")}
    xlsx<-createWorkbook()
    addWorksheet(xlsx,Sheet1)
    writeDataTable(xlsx,Sheet1,DEGs_res_out,startCol = 1,startRow = 1)
    freezePane(xlsx, Sheet1, firstActiveRow = 1, firstActiveCol = 5, firstRow = FALSE, firstCol = FALSE)
    saveWorkbook(
      xlsx, 
      file = file.path(
        WD,"02.TableOutput", 
        paste0("STAR_DEGs_ResOut_",sprintf("%02d", i),"_",gsub(" ", "", Par_ConLs_C[i]),".xlsx")
      ),overwrite = TRUE
    )
    rm(xlsx)
    #|----All enrichment resuts output (GSEA)----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Outputing GSEA results")}
    xlsx<-createWorkbook()
    for (j in 1:4) {
      if(nrow(as.data.frame(DEGs_Div$GSEA_Enrich[[j]]$Sig))>0){
        addWorksheet(xlsx,c("GO_BP","GO_MF","GO_CC","KEGG")[j])
        out_convert<-DEGs_Div$GSEA_Enrich[[j]]$Sig
        if(keyType!="ENSEMBL"){
          for(rn in 1:nrow(out_convert)){
            entrez_id2<-ID_mapping$external_gene_name
            names(entrez_id2)<-ID_mapping$entrezgene_id
            rn_tmp<-out_convert$core_enrichment[rn]%>%
              str_split(.,"/")%>%.[[1]]%>%entrez_id2[.]
            rn_tmp[rn_tmp%in%""]<-out_convert$core_enrichment[rn]%>%
              str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
            out_convert$core_enrichment[rn]<-rn_tmp%>%paste(.,collapse="||")
          }
        }else{
          for(rn in 1:nrow(out_convert)){
            if(j==4){
              entrez_id2<-ID_mapping$external_gene_name
              names(entrez_id2)<-ID_mapping$entrezgene_id
              rn_tmp<-out_convert$core_enrichment[rn]%>%
                str_split(.,"/")%>%.[[1]]%>%entrez_id2[.]
              rn_tmp[rn_tmp%in%""]<-out_convert$core_enrichment[rn]%>%
                str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
              out_convert$core_enrichment[rn]<-rn_tmp%>%paste(.,collapse="||")
            }else{
              ensembl_id<-ID_mapping$external_gene_name
              names(ensembl_id)<-ID_mapping$ensembl_gene_id
              rn_tmp<-out_convert$core_enrichment[rn]%>%
                str_split(.,"/")%>%.[[1]]%>%ensembl_id[.]
              rn_tmp[rn_tmp%in%""]<-out_convert$core_enrichment[rn]%>%
                str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
              out_convert$core_enrichment[rn]<-rn_tmp%>%paste(.,collapse="||")
            }
          }
        }
        writeDataTable(xlsx,c("GO_BP","GO_MF","GO_CC","KEGG")[j],
                       out_convert,startCol = 1,startRow = 1)
      }
    }
    if(length(names(xlsx))>0){
      for (k in 1:length(names(xlsx))) {
        setColWidths(xlsx, sheet=k, cols=1, widths = 12, hidden = FALSE, ignoreMergedCells = FALSE)
        setColWidths(xlsx, sheet=k, cols=2, widths = 45, hidden = FALSE, ignoreMergedCells = FALSE)
        freezePane(xlsx, sheet=k, firstActiveRow = 1, firstActiveCol = 3, firstRow = FALSE, firstCol = FALSE)
      }
    }else{
      addWorksheet(xlsx,"NULL")
    }
    saveWorkbook(
      xlsx, 
      file.path(
        WD,"02.TableOutput",
        paste0("GSEA_EnrichmentResults_",sprintf("%02d", i),"_",gsub(" ", "", Par_ConLs_C[i]),".xlsx")
      ),overwrite = TRUE
    )
    rm(xlsx)
    #|----All enrichment resuts output (Over Representation Test)----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Outputing ORA results")}
    for (l in c("All","Up_regulated","Down_regulated")) {
      xlsx<-createWorkbook()
      for (j in 1:4) {
        if(nrow(as.data.frame(DEGs_Div$Over_Represent[[l]][[j]]$Sig))>0){
          addWorksheet(xlsx,c("GO_BP","GO_MF","GO_CC","KEGG")[j])
          out_convert<-DEGs_Div$Over_Represent[[l]][[j]]$Sig
          if(keyType!="ENSEMBL"){
            for(rn in 1:nrow(out_convert)){
              entrez_id2<-ID_mapping$external_gene_name
              names(entrez_id2)<-ID_mapping$entrezgene_id
              rn_tmp<-out_convert$geneID[rn]%>%
                str_split(.,"/")%>%.[[1]]%>%entrez_id2[.]
              rn_tmp[rn_tmp%in%""]<-out_convert$geneID[rn]%>%
                str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
              out_convert$geneID[rn]<-rn_tmp%>%paste(.,collapse="||")
            }
          }else{
            for(rn in 1:nrow(out_convert)){
              if(j==4){
                entrez_id2<-ID_mapping$external_gene_name
                names(entrez_id2)<-ID_mapping$entrezgene_id
                rn_tmp<-out_convert$geneID[rn]%>%
                  str_split(.,"/")%>%.[[1]]%>%entrez_id2[.]
                rn_tmp[rn_tmp%in%""]<-out_convert$geneID[rn]%>%
                  str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
                out_convert$geneID[rn]<-rn_tmp%>%paste(.,collapse="||")
              }else{
                ensembl_id<-ID_mapping$external_gene_name
                names(ensembl_id)<-ID_mapping$ensembl_gene_id
                rn_tmp<-out_convert$geneID[rn]%>%
                  str_split(.,"/")%>%.[[1]]%>%ensembl_id[.]
                rn_tmp[rn_tmp%in%""]<-out_convert$geneID[rn]%>%
                  str_split(.,"/")%>%unlist()%>%.[rn_tmp%in%""]
                out_convert$geneID[rn]<-rn_tmp%>%paste(.,collapse="||")
              }
            }
          }
          writeDataTable(xlsx,c("GO_BP","GO_MF","GO_CC","KEGG")[j],
                         out_convert,startCol = 1,startRow = 1)
        }
      }
      if(length(names(xlsx))>0){
        for (k in 1:length(names(xlsx))) {
          setColWidths(xlsx, sheet=k, cols=1, widths = 12, hidden = FALSE, ignoreMergedCells = FALSE)
          setColWidths(xlsx, sheet=k, cols=2, widths = 45, hidden = FALSE, ignoreMergedCells = FALSE)
          freezePane(xlsx, sheet=k, firstActiveRow = 1, firstActiveCol = 3, firstRow = FALSE, firstCol = FALSE)
        }
      }else{
        addWorksheet(xlsx,"NULL")
      }
      saveWorkbook(
        xlsx, 
        file.path(
          WD,"02.TableOutput",
          paste0("OverRepresentTest_",sprintf("%02d", i),"_",gsub(" ", "", Par_ConLs_C[i]),"_",l,".xlsx")
        ),overwrite = TRUE
      )
      rm(xlsx)
    }
    #
    #|----Output----|####
    if(ShinyProgress){incProgress(0.05/length(Par_ConLs_C), detail = "Output finishing")}
    DEGs_ls[[i]]<-DEGs_Div
    #Output FEA comparsion
    if(is.null(DEGs_ls[[i]]$Over_Represent)){next()}
    temp<-DEGs_ls[[i]]$Over_Represent
    for(k in c("GO_BP","GO_MF","GO_CC","KEGG")){
      mydf<-data.frame()
      geneClusters<-list()
      for(j in c("All","Up_regulated","Down_regulated")){
        if(is.null(nrow(temp[[j]][[k]][["Sig"]]))){next()}
        mydf<-rbind(
          mydf,
          cbind(
            group=rep(j,nrow(temp[[j]][[k]][["Sig"]])),
            temp[[j]][[k]][["Sig"]]
          )
        )
        geneClusters[[j]]<-(temp[[j]][[k]][["Sig"]]$geneID)%>%
          str_split(.,"/")%>%unlist()%>%unique()
      }
      graphics.off()
      if(!rownames(mydf)%>%is_empty()){
        graphics.off()
        rownames(mydf)<-1:nrow(mydf)
        mydf$group<-factor(mydf$group,levels=c("All","Up_regulated","Down_regulated"))
        colnames(mydf)[1]<-"Cluster"
        res <- new("compareClusterResult",
                   compareClusterResult = mydf,
                   geneClusters = geneClusters
        )
        .call<-match.call(compareCluster,call("compareCluster",
                                              ENSEMBL~group, data=mydf, fun="enrichGO",
                                              OrgDb         = edb,
                                              keyType       = "ENSEMBL",
                                              ont           = "BP",
                                              pAdjustMethod = "BH",
                                              pvalueCutoff  = 1,
                                              qvalueCutoff  = 1,
                                              readable      = TRUE))
        .call$geneClusters<-as.symbol("geneClusters")
        .call$OrgDb<-as.symbol("edb")
        .call$data<-NULL
        res@.call<-.call
        res@fun<-"enrichGO"
        res@keytype<-"kegg"
        res@readable<-FALSE
        gp<-enrichplot::dotplot(res, showCategory=20, label_format=100, color = "pvalue",
                                title=paste0("Over Representation Test Summary of ",names(DEGs_ls)[i], " for ", k," Terms."))
        svg(
          file.path(WD,"01.PlotOutput",FolderName,"Over_Represent",paste0("00.Summary_",k,".svg")), 
          width = 13, height = 10, pointsize = 12
        )
        print(gp)
        dev.off()
      }
      graphics.off()
    }
    names(DEGs_ls)[i]<-gsub(" ", "", Par_ConLs_C[i])
    gc()
  }
  ColChk<-ColChk
  log2FColNa<-log2FColNa
  Group.By<-Group.By
  tbChangedFC<-1
  tbChangedGp<-1
  RunningData<-RunningData
  ID_mapping<-ID_mapping
  ColNaLs<-colnames(Data)
  tbOut<-tbOut
  DataChgFC<-Data
  DEGs_ls<-DEGs_ls
  Data<-Data
  DataChgGp<-tbOut
  EntrezID<-EntrezID
  save(
    ColChk, log2FColNa, Group.By,
    tbChangedFC, tbChangedGp, RunningData,
    ID_mapping, ColNaLs, tbOut, DataChgFC,
    DEGs_ls, Data, DataChgGp, EntrezID,
    file = file.path(WD,"TheLastRunResults.RData")
  )
  if(ShinyProgress){incProgress(0.2, detail = "Done")}
}
```

```{r 00-fin}
# Clean up
rm(
  u, url, attemp, attempt, attempt_max, list.of.packages.Bio, list.of.packages, new.packages, new.packages.Bio
)
gc()
```

### Make functions to plot distribution of features batchly.

```{r}
WZY_FeatureBlukPlot <- function(i, Data_temp, geneLs, fovs, OnlyCluster=FALSE){
  PlotLs <- list()
  SelectedCluster <- rep("Others", ncol(Data_temp))
  SelectedCluster[Data_temp$seurat_clusters==i] <- i
  SelectedCluster <- factor(SelectedCluster, levels = c("Others", i))
  names(SelectedCluster) <- colnames(Data_temp)
  Data_temp <- AddMetaData(Data_temp, metadata = SelectedCluster, col.name = "SelectedCluster")
  Idents(Data_temp) <- "SelectedCluster"
  for (loc in fovs) {
    Plot <- ImageDimPlot(
      Data_temp, cols = c("gray", "red"),
      fov=loc, alpha = 0.7,
      size = 1.5, dark.background = FALSE
    )+ggtitle(loc) + guides(fill=guide_legend(ncol=1))
    PlotLs <- c(PlotLs, list(Plot))
  }
  if(!OnlyCluster){
    for(gene in geneLs){
      for (loc in fovs) {
        Plot <- ImageFeaturePlot(
          Data_temp, features = gene, fov = loc, size = 1.2, alpha = 0.9,
          cols = c("gray90", "red"), dark.background = FALSE
        )
        PlotLs <- c(PlotLs, list(Plot))
      }
    }
  }
  return(PlotLs)
}
```

# 01.Nuclei Segmentation and Custom Binning of Visium HD Gene Expression Data

## 01-01.Jupyter Notebook

Tutorial is from 10x Genomics: https://www.10xgenomics.com/analysis-guides/segmentation-visium-hd

Because the data is huge and Rstudio running with such huge data in python chunk is super slow, all the python codes were running inside a jupyter notebook.

```{bash}
# To enable TOC inside jupyter notebook:
## pip install --upgrade --no-cache-dir notebook==6.*
## pip install jupyter_contrib_nbextensions
## jupyter contrib nbextension install --user
## jupyter nbextension enable codefolding/main
## jupyter nbextension enable toc2/main
## pip install --upgrade --no-cache-dir notebook==7
conda activate $HOME/.local/share/r-miniconda/envs/stardist_env
jupyter lab 01.scRNAseq10xVisiumHD.ipynb
```

Click [HERE](./01.scRNAseq10xVisiumHD.html) to check the output of the above Jupyter Notebook. 

## 01-02.Convert AnnData to Seurat object

From the above Jupyter Notebook, the centroids and boundaries of cells, the coordinates of genes, and the gene expression matrix were exported as `centroid_df.txt`, `boundary_df.txt`, `merged_df.txt`, and `count_area_filtered_adata.h5`, respectively. Then the next, we need to use `SeuratObject` to construct a `Seurat` object as *Image-based Spatial Data (Xenium)*.

```{r 01-02-1}
# install dependencies
if(!require("hdf5r")){install.packages("hdf5r")}
if(!require("Seurat")){install.packages("Seurat")}
if(!require("SeuratObject")){install.packages("SeuratObject")}
if(!require("SeuratDisk")){remotes::install_github("mojaveazure/seurat-disk")}
# install platform specific source package
if (!require("loupeR")){
  os <- sub("Darwin", "macOS", Sys.info()["sysname"])
  url <- paste0("https://github.com/10XGenomics/loupeR/releases/latest/download/loupeR_", os, ".tar.gz")
  install.packages(url, repos = NULL, type = "source")
  loupeR::setup()
}
library(hdf5r)
library(Seurat)
library(SeuratObject)
library(SeuratDisk)
library(loupeR)
```

Import the centroids and boundaries of cells, the coordinates of genes, and the gene expression matrix. Note that: cells with no genes and genes with no cells need to be removed. 

```{r 01-02-2}
# Set data type
assay <- "Xenium"
fov <- "fov"
# Load ene expression matrix
matrix <- Read10X_h5(filename="./count_area_filtered_adata.h5", use.names = TRUE, unique.features = TRUE)
## Remove genes expressed less than 6 cells
GeneKeep <- rownames(matrix)[(rowSums(matrix) > 6)]
matrix <- matrix[GeneKeep,]
## Remove cells expressed less than 6 genes
CellKeep <- colnames(matrix)[(colSums(matrix) > 6)]
matrix <- matrix[, CellKeep]
# Load centroids and boundaries
centroids <- read.table(file = "./centroid_df.txt", sep = "\t", header = TRUE)
colnames(centroids) <- c("cell", "x", "y")
centroids <- centroids[,c("x", "y", "cell")]
segmentations <- read.table(file = "./boundary_df.txt", sep = "\t", header = TRUE)
colnames(segmentations) <- c("cell", "x", "y")
# Load gene coordinates
microns <- read.table(file = "./merged_df.txt", sep = "\t", header = TRUE)
microns <- microns[, c("x", "y", "gene")]
```

Check the data and remove cells and genes that are not present in the matrix.

```{r 01-02-3}
# Extract cell IDs from centroids, segmentations, and microns
centroid_cells <- centroids$cell
segmentation_cells <- segmentations$cell
# Extract cell IDs from the matrix
matrix_cells <- colnames(matrix)
# Check for mismatches
setdiff(centroid_cells, matrix_cells) %>% head()         # Cells in centroids not present in matrix
setdiff(segmentation_cells, matrix_cells) %>% head()     # Cells in segmentations not present in matrix
# Filter the centroids and segmentations to include only cells present in the matrix
common_cells <- intersect(centroid_cells, matrix_cells)
matrix <- matrix[,common_cells]
GeneKeep <- rownames(matrix)[(rowSums(matrix) > 6)]
matrix <- matrix[GeneKeep,]
CellKeep <- colnames(matrix)[(colSums(matrix) > 6)]
matrix <- matrix[, CellKeep]
centroids <- centroids[centroids$cell %in% common_cells, ]
segmentations <- segmentations[segmentations$cell %in% common_cells, ]
microns <- microns[microns$gene %in% GeneKeep, ]
# Reorder centroids and segmentations to match the order in the matrix
centroids <- centroids[order(match(centroids$cell, common_cells)), ]
segmentations <- segmentations[order(match(segmentations$cell, common_cells)), ]
```

In this time (2024-10-08), the 10x Visium, Xenium, and Visium HD is not supported by the LoupeR. Therefore, we shall convert all barcodes to the 10x Genomics 3 Single Cell Gene Expression dataset. The [10x barcode whitelist](https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist) (*3M-february-2018.txt*) is avaiable in `cellranger-x.y.z/lib/python/cellranger/barcodes/`, if the [`cellranger`](https://www.10xgenomics.com/support/software/cell-ranger/latest) is installed.

Please note that if the barcodes come from non-10x assays, this is not supported by LoupeR, since its license is only for use with 10x Genomics experiments.

```{r 01-02-4}
# Convert barcodes to 10x Genomics 3' Single Cell Gene Expression dataset
num_cells <- nrow(centroids)
set.seed(1453)
new_barcodes <- read.table(file = "./3M-february-2018.txt", sep = "\t", header = FALSE)
new_barcodes <- new_barcodes$V1
if(num_cells < length(new_barcodes)){
  # If there are fewer cells than barcodes, sample the barcodes
  new_barcodes <- sample(new_barcodes, num_cells)
  new_barcodes <- paste0(new_barcodes,"-1")
}else{
  # If there are more cells than barcodes, sample with replacement
  new_barcodes <- sample(new_barcodes, num_cells, replace = TRUE)
  new_barcodes <- paste0(new_barcodes,"-1")
  new_barcodes <- unique(new_barcodes)
  BarcodesNeed <- num_cells - length(new_barcodes)
  pb = txtProgressBar(min = 1, max = BarcodesNeed, initial = 1)
  while (length(new_barcodes) < num_cells) {
    new_barcodes <- unique(
      c(
        new_barcodes,
        replicate(
          num_cells - length(new_barcodes),
          paste0(
            sample(new_barcodes, 1), "-", round(runif(1, min = 0, max = 9)), round(runif(1, min = 0, max = 9))
          )
        )
      )
    )
    setTxtProgressBar(pb, BarcodesNeed - (num_cells - length(new_barcodes)))
  }
  close(pb)
}
# Validate the barcodes
loupeR:::validate_barcodes(barcodes=new_barcodes)
# Assign the new barcodes to the counts matrix
names(new_barcodes) <- centroids$cell
centroids$cell <- new_barcodes[centroids$cell] %>% as.vector()
segmentations$cell <- new_barcodes[segmentations$cell] %>% as.vector()
colnames(matrix) <- new_barcodes[colnames(matrix)] %>% as.vector()
```

Now, we can create the `SeuratObject` using the `SeuratObject` package.

```{r 01-02-5}
# Parameters setting
type <- c("centroids", "segmentations")
outs <- c("matrix", "microns")
outs <- c(outs, type)
has_dt <- requireNamespace("data.table", quietly = TRUE) && requireNamespace("R.utils", quietly = TRUE)
# Create 10x-at objectke data
data <- sapply(outs, function(otype) {
  switch(
    EXPR = otype,
    'matrix' = {
      matrix
    },
    'centroids' = {
      centroids
    },
    'segmentations' = {
      segmentations
    },
    'microns' = {
      microns
    },
    stop("Unknown Xenium input type: ", otype)
  )
}, USE.NAMES = TRUE)
# Create 10x-Xenium-like data
segmentations.data <- list(
  "centroids" = CreateCentroids(data$centroids),
  "segmentation" = CreateSegmentation(data$segmentations)
)
coords <- CreateFOV(
  coords = segmentations.data,
  type = c("segmentation", "centroids"),
  molecules = data$microns,
  assay = assay
)
# Create Seurat object
xenium.obj <- CreateSeuratObject(counts = data$matrix, assay = assay)
xenium.obj@images[[fov]] <- coords
```

## 01-03.Generate CLOUPE files from Seurat Objects

[Routine pre-processing of Seurat objects](https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2)

```{r 01-03-1}
xenium.obj <- NormalizeData(xenium.obj)
xenium.obj <- ScaleData(xenium.obj)
xenium.obj <- FindVariableFeatures(xenium.obj, selection.method = "vst", nfeatures = 2000)
xenium.obj <- SCTransform(xenium.obj, assay = "Xenium", clip.range = c(-10, 10))
xenium.obj <- RunPCA(xenium.obj, npcs = 30, features = rownames(xenium.obj))
xenium.obj <- RunUMAP(xenium.obj, dims = 1:30)
xenium.obj <- FindNeighbors(xenium.obj, reduction = "pca", dims = 1:30)
xenium.obj <- FindClusters(xenium.obj, resolution = 1.2)
```

Check the clustering results.

```{r 01-03-2}
DefaultBoundary(xenium.obj.crop[["fov"]]) <- "segmentation"
ImageDimPlot(xenium.obj, fov = "fov", cols = "polychrome", axes = TRUE)
```

Now, we can generate the Loupe file from the Seurat object.

```{r 01-03-3}
# Gene Expression RNA assay
assay <- xenium.obj[["Xenium"]]
# get counts matrix from either the old or newer formats of assay
counts <- counts_matrix_from_assay(assay)
# get clusters from the Seurat object
clusters <- select_clusters(xenium.obj)
# get projections from the Seurat object
projections <- select_projections(xenium.obj)
# get morphologies from the Seurat object
morph <- xenium.obj@images[["fov"]]@boundaries[["centroids"]]@coords %>% as.matrix()
rownames(morph) <- xenium.obj@images[["fov"]]@boundaries[["centroids"]]@cells
# add the morphologies to the projections
projections <- c(projections, morph = list(morph))
# get the barcodes from the counts matrix
barcodes <- colnames(counts)
barcode_count <- length(barcodes)
# validate the data
loupeR:::validate_count_mat(counts)
loupeR:::validate_clusters(clusters, barcode_count)
loupeR:::validate_projections(projections, barcode_count)
# convert the count matrix, clusters, and projections into a Loupe file
if(!dir.exists("./00.WZY_Res")){
  dir.create("./00.WZY_Res")
}
create_loupe(
  counts,
  clusters = clusters,
  projections = projections,
  force = TRUE,
  output_dir = "./00.WZY_Res",
  output_name = "WZY_Res",
  feature_ids = NULL,
  executable_path = NULL,
  seurat_obj_version = NULL
)
```

Finishing and clean

```{r 01-fin-1}
# Save the Seurat object
saveRDS(xenium.obj, file = "xenium.obj.rds")
# Clean up
rm(
  # Environments variables
  u, url, attemp, attempt, attempt_max, list.of.packages.Bio, list.of.packages, new.packages, new.packages.Bio,
  # Exported AnnData files from Jupyter Notebook
  centroids, matrix, microns, segmentations, CellKeep, GeneKeep, centroid_cells, matrix_cells, common_cells,
  # For Seurat Objects
  data, segmentations.data, segmentation_cells, type, fov, has_dt, outs, num_cells,
  # For LoupeR
  assay, barcode_count, barcodes, BarcodesNeed, new_barcodes, pb, morph, projections, clusters, coords, counts
)
```

```{r 01-fin-2}
rm(xenium.obj)
gc()
save.image(file = "BreakPoint.RData")
```

# 02.Further analysis on specific ROIs

Clustering a subset of the single-cell RNA sequencing (scRNA-seq) dataset can potentially lead to better results, as suggested by the following findings:

- 1. [Zhu X, Wang J, Li R, Peng X. Comparison of Gene Selection Methods for Clustering Single-cell RNA-seq Data. Current Bioinformatics. 2023 Jan 1;18(1):1-1.](https://www.ingentaconnect.com/content/ben/cbio/2023/00000018/00000001/art00002)
- 2. [Renaut S, Saavedra Armero V, Boudreau DK, Gaudreault N, Desmeules P, Thriault S, Mathieu P, Joubert P, Boss Y. Single-cell and single-nucleus RNA-sequencing from paired normal-adenocarcinoma lung samples provide both common and discordant biological insights. PLoS Genetics. 2024 May 30;20(5):e1011301.](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1011301)

The study from Zhu X et al. (2023) compared different gene selection methods for clustering scRNA-seq data, including methods that leverage subsets of the data. The authors found that methods that select genes based on the expression variance across a subset of cells (e.g. top variable genes) can outperform methods that use the full dataset. Specifically, they showed that using a subset of cells for gene selection and then applying clustering on the full dataset can lead to better clustering performance compared to using the full dataset for both gene selection and clustering. The authors suggest that this is because scRNA-seq data is high-dimensional, and focusing on a subset of informative genes/cells can help overcome the curse of dimensionality.

The study from Renaut S et al. (2024) analyzed scRNA-seq data from paired normal and tumor lung samples to identify cell type-specific gene expression changes. The authors note that cell type annotation methods for scRNA-seq data have been largely built from scRNA-seq datasets, and therefore, analyzing a subset of the data can provide both common and discordant biological insights. By focusing the analysis on specific cell types or subsets of the data, the authors were able to uncover novel cell type-specific gene expression changes that were missed when analyzing the full dataset. The authors suggest that this subset-based approach can be valuable for identifying rare cell types or subtle expression changes that may be obscured when considering the entire scRNA-seq dataset.

In summary, the above studies indicate that clustering a subset of the scRNA-seq dataset can lead to better results by addressing the high dimensionality of the data and providing more targeted insights, especially for identifying rare cell types or subtle expression changes.

## 02-01.Load Data & Environment Setting

```{r 02-01-1}
load("./BreakPoint.RData")
xenium.obj <- readRDS(file = "./xenium.obj.rds")
CellsInROIs1 <- read.csv(file = "E12.5.csv", header = TRUE)
CellsInROIs2 <- read.csv(file = "E13.5.csv", header = TRUE)
```

## 02-02.Crop data into ROIs {.tabset .tabset-fade}

### 1.Get Cell List

```{r 02-02-1}
# Define the cells to keep for each ROI
CellKept <- list(
  E12_5_AB_KO.CellKept = CellsInROIs1$Barcode[CellsInROIs1$E12.5 %in% "Palate_KO"],
  E12_5_AB_WT.CellKept = CellsInROIs1$Barcode[CellsInROIs1$E12.5 %in% "Palate_WT"],
  E13_5_AB_KO.CellKept = CellsInROIs2$Barcode[CellsInROIs2$E13.5 %in% "Palate_KO"],
  E13_5_AB_WT.CellKept = CellsInROIs2$Barcode[CellsInROIs2$E13.5 %in% "Palate_WT"]
)
xenium.obj.crop <- subset(xenium.obj, cells = as.vector(unlist(CellKept)))
# Define the FOVs
FOVs <- c("E12.5_AB_KO", "E12.5_AB_WT", "E13.5_AB_KO", "E13.5_AB_WT")
for (FOVs.idx in 1:length(CellKept)) {
  Cell.ls <- CellKept[[FOVs.idx]]
  Cell.Coords <- cbind(
    CellID = xenium.obj.crop@images[[1]]$centroids@cells,
    x = xenium.obj.crop@images[[1]]$centroids@coords[, "x"],
    y = xenium.obj.crop@images[[1]]$centroids@coords[, "y"]
  ) %>% as.data.frame()
  Cell.Coords$x <- Cell.Coords$x %>% as.numeric()
  Cell.Coords$y <- Cell.Coords$y %>% as.numeric()
  rownames(Cell.Coords) <- Cell.Coords$CellID
  # Crop the data
  xenium.obj.crop[[FOVs[FOVs.idx]]] <- Crop(
    xenium.obj.crop@images[[1]], coords = "tissue",
    x = c(min(Cell.Coords[Cell.ls,"x"], na.rm = TRUE), max(Cell.Coords[Cell.ls,"x"], na.rm = TRUE)),
    y = c(min(Cell.Coords[Cell.ls,"y"], na.rm = TRUE), max(Cell.Coords[Cell.ls,"y"], na.rm = TRUE))
  )
  DefaultBoundary(xenium.obj.crop[[FOVs[FOVs.idx]]]) <- "segmentation"
}
```

### 2.Check ROIs

```{r 02-02-2}
# Check the cropped data
Plot.ls <- list()
for (FOVs.idx in 1:length(CellKept)) {
  Plot <- ImageDimPlot(
    xenium.obj.crop, cols = "polychrome",
    fov=FOVs[FOVs.idx],
    size = 2, dark.background = F
  ) + theme(legend.position="none", plot.title = element_text(size=18)) + ggtitle(FOVs[FOVs.idx])
  Plot.ls <- c(Plot.ls, list(Plot))
}
```

```{r 02-02-3, fig.width = 16, fig.height = 8}
# Display the cropped data
par(mar = c(0,0,0,0), xpd=NA)
grid.arrange(grobs = Plot.ls, ncol = 2)
```

### 3.QC {.tabset .tabset-dropdown}

Quality control to remove cells expressing too much mtRNA and too few genes.

```{r}
# Load the cropped data
Treatment <- "PAX9_KO"
# Get cell groups
Cell.ls <- unlist(CellKept)
Cell.ls <- Cell.ls[order(match(Cell.ls, colnames(xenium.obj.crop)))]
# Add metadata
condition <- rep("WT", ncol(xenium.obj.crop))
condition[grepl("KO", names(Cell.ls))] <- Treatment
xenium.obj.crop <- AddMetaData(xenium.obj.crop, condition, "condition")
# Calculate the percentage of mtRNA for each cell
xenium.obj.crop[["percent.mt"]] <- PercentageFeatureSet(xenium.obj.crop, pattern = "^mt-", assay = "Xenium")
```

#### Before QC

```{r, fig.width=15, fig.height=10}
# Visualize QC metrics as a violin plot
VlnPlot(xenium.obj.crop, features = c("nFeature_Xenium", "nCount_Xenium", "percent.mt"), ncol = 3, group.by="condition")
```

```{r, fig.width=15, fig.height=7}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot1 <- FeatureScatter(xenium.obj.crop, feature1 = "nCount_Xenium", feature2 = "percent.mt", group.by="condition")
plot2 <- FeatureScatter(xenium.obj.crop, feature1 = "nCount_Xenium", feature2 = "nFeature_Xenium", group.by="condition")
plot1 + plot2
```

#### QC

- We filter cells that have unique feature counts less than 25
- We filter cells that have >10% mitochondrial counts

```{r}
xenium.obj.crop <- subset(xenium.obj.crop, subset = nFeature_Xenium > 25 & percent.mt < 10)
```

#### After QC

```{r, fig.width=15, fig.height=10}
# Visualize QC metrics as a violin plot
VlnPlot(xenium.obj.crop, features = c("nFeature_Xenium", "nCount_Xenium", "percent.mt"), ncol = 3, group.by="condition")
```

```{r, fig.width=15, fig.height=7}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot1 <- FeatureScatter(xenium.obj.crop, feature1 = "nCount_Xenium", feature2 = "percent.mt", group.by="condition")
plot2 <- FeatureScatter(xenium.obj.crop, feature1 = "nCount_Xenium", feature2 = "nFeature_Xenium", group.by="condition")
plot1 + plot2
```

### 3.Finishing

Save current results and cleanup for the next step

```{r 02-02-fin-1}
# Save global variables
save(
  CellKept, project, FOVs,
  file = "GlobalVariables.RData"
)
```

```{r 02-02-fin-2}
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
# Clean up
rm(
  # Data
  CellsInROIs1, CellsInROIs2,
  # Ploting
  Plot, Plot.ls, Cell.ls, FOVs.idx, Cell.Coords,
  # Seurat Objects
  xenium.obj, xenium.obj.crop
)
gc()
save.image(file = "BreakPoint.RData")
```

## 02-03.Clustering {.tabset .tabset-fade}

```{r 02-03-0}
load("GlobalVariables.RData")
```

### Clustering & Analysis {.tabset .tabset-dropdown}

#### Re-Clustering

Do re-clustering with cropped ROIs.

```{r 02-03-1}
# Load the cropped data
Treatment <- "PAX9_KO"
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
# Get cell groups
Cell.ls <- unlist(CellKept)
Cell.ls <- Cell.ls[order(match(Cell.ls, colnames(xenium.obj.crop)))]
# Add metadata
condition <- rep("WT", ncol(xenium.obj.crop))
condition[grepl("KO", names(Cell.ls))] <- Treatment
xenium.obj.crop <- AddMetaData(xenium.obj.crop, condition, "condition")
age <- rep("placebo", ncol(xenium.obj.crop))
age[grepl("E12_5", names(Cell.ls))] <- "E12.5"
age[grepl("E13_5", names(Cell.ls))] <- "E13.5"
xenium.obj.crop <- AddMetaData(xenium.obj.crop, age, "age")
orig.ident <- rep("placebo", ncol(xenium.obj.crop))
orig.ident[grepl("E12_5_AB_KO.CellKept", names(Cell.ls))] <- "E12.5_PAX9_KO"
orig.ident[grepl("E12_5_AB_WT.CellKept", names(Cell.ls))] <- "E12.5_WT"
orig.ident[grepl("E13_5_AB_KO.CellKept", names(Cell.ls))] <- "E13.5_PAX9_KO"
orig.ident[grepl("E13_5_AB_WT.CellKept", names(Cell.ls))] <- "E13.5_WT"
xenium.obj.crop <- AddMetaData(xenium.obj.crop, orig.ident, "orig.ident")
# Re-clustering
xenium.obj.crop <- FindVariableFeatures(xenium.obj.crop, selection.method="vst", nfeatures = 1000)
dims <- round(min(ncol(xenium.obj.crop),nrow(xenium.obj.crop))/3)
features <- VariableFeatures(xenium.obj.crop)
xenium.obj.crop <- SCTransform(xenium.obj.crop, assay = "Xenium") %>%
  RunPCA(., npcs = dims, features = features)
Cutdim <- xenium.obj.crop@reductions[["pca"]]@stdev%>%
  .[.>quantile(., .95)]%>%length()
xenium.obj.crop <- RunUMAP(xenium.obj.crop, dims = 1:Cutdim, min.dist = 1, n.neighbors = 12, n.epochs = 500)
xenium.obj.crop <- FindNeighbors(xenium.obj.crop, reduction = "pca", dims = 1:Cutdim)
xenium.obj.crop <- FindClusters(xenium.obj.crop, resolution = 0.6)
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
```

#### Find Cell Markers

Identifies 'markers' of gene expression using ROC analysis.

```{r 02-03-2}
xenium.obj.crop <- PrepSCTFindMarkers(xenium.obj.crop)
Res.Cell_Markers <- FindAllMarkers(xenium.obj.crop, test.use = "roc", return.thresh = 0, logfc.threshold = 0)
xenium.obj.crop@misc[["Cell.Markers"]] <- Res.Cell_Markers
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
```

#### Get gene annotations

```{r 02-03-3}
GeneAnno <- GeneName2ID_PubMedBioMat_UniProt(
  genels = rownames(xenium.obj.crop), OrgList = OrgList, species = "Mouse genes (GRCm39)",
  dataset = OrgList["Mouse genes (GRCm39)", "dataset"], ShinyProgress = FALSE, ShinyProg.N = 1
)
ID_mapping <- GeneAnno$ID_mapping
EntrezID <- GeneAnno$EntrezID
save(
  GeneAnno, ID_mapping, EntrezID, file=file.path("Genes_IDConvert.RData")
)
```

#### Do FEA Analysis

Do Functional Enrichment Analysis (FEA).

```{r 02-03-4, warning=FALSE}
FEA_ResNa <- "Xenium.obj.crop_FEA_CMarkers"
load("Genes_IDConvert.RData")
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
temp <- ID_mapping
rownames(temp) <- make.unique(temp$GeneSymbol)
Data <- cbind(
  xenium.obj.crop@misc[["Cell.Markers"]],
  temp[xenium.obj.crop@misc[["Cell.Markers"]]$gene,]
)
Data$GeneSymbol <- Data$gene
Data <- Data[,-8]
WZY_RunFEA(
  log2FColNa = "avg_log2FC", ColChk = TRUE, Group.By = "cluster", Data = Data,
  tbOut = Data, WD = WD, FilePath = FEA_ResNa, ShinyProgress = FALSE,
  AnnoHub = "org.Mm.eg.db", Table1_rows_all = rownames(Data)[Data$power>0.4],
  Par_UseFDR4FEA = FALSE, Par_FDR4FEA = 0.05, Par_FDR4KEG_OR = 0.05,
  Par_species.go.kegg = "mmu", descend = TRUE
)
```

Zip FEA output files for [Download](./Xenium.obj.crop_FEA_CMar_FEA_Res.zip)

```{bash, engine.opts='-l'}
zip -r ./Xenium.obj.crop_FEA_CMar_FEA_Res.zip ./Xenium.obj.crop_FEA_CMar_FEA_Res
```

### Seurat Clusters

```{r 02-01-2.a, fig.width=16, fig.height=8}
cols <- toupper(as.vector(dark.colors(24)))
names(cols) <- Idents(xenium.obj.crop)%>% unique() %>% mixedsort() %>% as.vector()
Plot.ls <- list()
for (FOVs.idx in 1:length(CellKept)) {
  Plot <- ImageDimPlot(
    xenium.obj.crop, cols = cols,
    fov=FOVs[FOVs.idx],
    size = 2, dark.background = F
  ) + guides(fill=guide_legend(ncol=2)) + theme(plot.title = element_text(size=18)) + ggtitle(FOVs[FOVs.idx])
  Plot.ls <- c(Plot.ls, list(Plot))
}
# Display the cropped data
par(mar = c(0,0,0,0), xpd=NA)
do.call("grid.arrange", c(Plot.ls, ncol=2))
```

```{r 02-01-2.b, fig.width=16, fig.height=14}
pt.size <- 0.8
alpha <- 0.5
PlotLs <- list(
  DimPlot(xenium.obj.crop, group.by = "orig.ident", pt.size = pt.size, alpha = 0.5) + ggtitle("orig.ident"),
  DimPlot(xenium.obj.crop, cols = cols, group.by = "seurat_clusters", pt.size = pt.size, alpha = 0.5) + ggtitle("seurat_clusters"),
  DimPlot(xenium.obj.crop, group.by = "condition", pt.size = pt.size, alpha = 0.5) + ggtitle("condition"),
  DimPlot(xenium.obj.crop, group.by = "age", pt.size = pt.size, alpha = 0.5) + ggtitle("age")
)
do.call("grid.arrange", c(PlotLs, ncol=2))
```

### Cell Markers

Identifies 'markers' of gene expression using ROC analysis.

For each gene, evaluates (using AUC, denote as "myAUC" in the table below) a classifier built on that gene alone, to classify between two groups of cells (target seurat_cluster versus all other seurat_clusters). An AUC value of 1 means that expression values for this gene alone can perfectly classify the two groupings (i.e. Each of the cells in cells.1 exhibit a higher level than each of the cells in cells.2). An AUC value of 0 also means there is perfect classification, but in the other direction.

A value of 0.5 implies that the gene has no predictive power to classify the two groups. Returns a 'power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially expressed genes.

pct.1: percentage of cells (in the target seurat_cluster) expressed this gene.

pct.2: percentage of cells (in all other seurat_clusters) expressed this gene.

```{r 02-01-3.a, echo=FALSE, warning=FALSE, cache=FALSE, results='asis'}
TB <- xenium.obj.crop@misc[["Cell.Markers"]]
write_xlsx(TB, "Xenium.obj.crop_Cell_Markers.xlsx")
TB <- TB[TB$power>0.4,]
tableID<-"Cell_Markers"
toggleID<-paste0(tableID,"FT")
colnm<-c(
  'myAUC', 'avg_diff', 'power', 'avg_log2FC',
  'pct.1', 'pct.2', 'cluster', 'gene'
)
TableResHTML.head<-hmakeTag(
  'th',colnm,
  style="font-weight: 900; border-bottom: 1px solid grey; border-top: 2px solid grey; text-align: center;"
  
)%>%
  paste(.,collapse="")%>%
  hmakeTag('tr',.)%>%hmakeTag('thead',.)
TableResHTML.body<-TB%>%
  hwrite(.,row.names=FALSE,col.names=FALSE,br=TRUE)%>%
  HTML()%>%read_html()%>%
  xml_find_all(.,".//tr")%>%as.character()%>%paste(.,collapse="")%>%
  hmakeTag('tbody',.)
TableResHTML<-hmakeTag('table',paste0(TableResHTML.head,TableResHTML.body))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//table")
ToggleBt<-hmakeTag('div', id=toggleID)
xml_set_attrs(TableResHTML[[1]], c("id"=tableID, "cellpadding"="0"))
JS_HTML_src<-hmakeTag('script', src='./HTML_source/tablefilter/tablefilter.js')
JS_HTML<-hmakeTag(
  'script', id='data-config',
  data = paste0(
    "
    var filtersConfig = {base_path: './HTML_source/tablefilter/', popup_filters: false, alternate_rows: true,
        msg_filter: 'Filtering...', rows_counter: true, btn_reset: true, status_bar: true,
        mark_active_columns: true, highlight_keywords: true, enable_default_theme: true,
        col_6: 'select',
        col_types: [
    ",
    paste(
      c('number', 'number', 'number', 'number', 'number',
        'number', 'number', 'string'), collapse="','"
    )%>%paste0("'",.,"'"),
    "
        ],
        extensions:[{name: 'sort' },
        {name: 'filtersVisibility', target_id: '",toggleID,"', btn_html: '<button>Toggle filters</button>'},
        {name: 'colsVisibility',
        text: 'Columns: ',tick_to_hide: false,enable_tick_all: true,
        headers_text: [
    ",
    paste(
      colnm, collapse="','"
    )%>%paste0("'",.,"'"),
    "
          ],
          at_start: []
        }],
        paging: {
          results_per_page: ['Records: ', [10, 25, 50, 100]]
        }
    };
    var tf = new TableFilter(",
    tableID%>%paste0("'",.,"'"),
    ", filtersConfig);
    tf.init();
    "
  )
)
Body_HTML<-hmakeTag('body',paste0(ToggleBt,TableResHTML,JS_HTML_src,JS_HTML))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//body")
as.character(Body_HTML)%>%HTML()
```

Click [HERE](./Xenium.obj.crop_Cell_Markers.xlsx) to download the above results as a .xlsx file.

```{r 02-01-3.b, fig.width=18, fig.height=9, message=FALSE}
df <- TB[TB$power>0.4,]
df <- df[df$avg_log2FC>0,]
DP <- DotPlot(
  object = xenium.obj.crop,
  features = unique(df$gene),
  assay = "SCT",
  scale = TRUE,
  dot.scale = 10
) + RotatedAxis() +
  scale_colour_gradientn(colours = c("#000000", "#FF0000")) +
  labs(title = "Cell Markers") +
  theme(text = element_text(size = 15))
print(DP)
```

### Distribution {.tabset .tabset-dropdown}

```{r}
TB <- xenium.obj.crop@misc[["Cell.Markers"]]
df <- TB[TB$power>0.4,]
df <- df[df$avg_log2FC>0,]
# Set plot parameters:
RowScaleFactor <- 8
PlotWidth <- 15
ncol <- 2
```

The "No significant cell markers." in this section means that there is no significant cell markers with log2FC > 0.

#### Seurat_Cluster 0

```{r}
i<-0
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=8, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 1

```{r}
i<-1
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=24, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 2

```{r}
i<-2
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=8, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 3

```{r}
i<-3
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=8, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 4

```{r}
i<-4
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=40, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 5

```{r}
i<-5
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=36, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 6

```{r}
i<-6
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=32, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 7

```{r}
i<-7
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=24, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 8

```{r}
i<-8
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=32, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 9

```{r}
i<-9
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=24, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 10

```{r}
i<-10
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=32, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 11

```{r}
i<-11
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=136, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

#### Seurat_Cluster 12

```{r}
i<-12
PlotHeight <- (length(df$gene[df$cluster==i])+1)*RowScaleFactor
print(PlotHeight)
```

```{r fig.height=40, fig.width=15}
if(PlotHeight>RowScaleFactor){
  geneLs <- df$gene[df$cluster==i]
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=FALSE), ncol=2
    )
  )
}else{
  do.call(
    "grid.arrange", c(
      WZY_FeatureBlukPlot(i = i, Data_temp = xenium.obj.crop, geneLs = geneLs, fovs = FOVs, OnlyCluster=TRUE), ncol=2,
      top = "No significant cell markers."
    )
  )
}
```

### FEA {.tabset .tabset-dropdown}

Click [HERE](./Xenium.obj.crop_FEA_CMar_FEA_Res.zip) to download all FEA results. The MS .xlsx files of all FEA results are avalible inside the downloaded .zip file.

```{r}
FEA_ResNa <- "Xenium.obj.crop_FEA_CMar"
```

#### Seurat_Cluster 0

```{r, fig.width=24, fig.height=10}
i<-1
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 1

```{r, fig.width=24, fig.height=10}
i<-2
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 2

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 3

```{r, fig.width=24, fig.height=10}
i<-4
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 4

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 5

```{r, fig.width=24, fig.height=10}
i<-4
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 6

```{r, fig.width=24, fig.height=10}
i<-5
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 7

```{r, fig.width=24, fig.height=10}
i<-6
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 8

```{r, fig.width=24, fig.height=10}
i<-7
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 9

```{r, fig.width=24, fig.height=10}
i<-9
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 10

```{r, fig.width=24, fig.height=10}
i<-9
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 11

```{r, fig.width=24, fig.height=10}
i<-10
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 12

```{r, fig.width=24, fig.height=10}
i<-11
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "Up_regulated_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant cell markers with log2FC>0.")
  FilePath <- file.path(
    ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
    paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
  )
  if(!file.exists(FilePath)){
    print("There is no significant cell markers.")
  }else{
    fig_svg<-magick::image_read_svg(FilePath)
    par(mar = c(0,0,0,0), xpd=NA)
    plot(fig_svg)
  }
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

### Help {.tabset .tabset-dropdown}

#### About Cell Markers

Below is a demo to show how to check the Cell Markers results using our software:

<video width="1200" height="675" controls>
  <source src="CMarkers_Demo.mp4" type="video/mp4">
</video>

#### About FEA

Below is a demo to show how to check the FEA results using our software:

<video width="1200" height="675" controls>
  <source src="FEA_Res_Demo.mp4" type="video/mp4">
</video>

#### Miscellaneous

Save current results and cleanup for the next step

```{r 02-03-fin}
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
# Clean up
rm(
  # Gene Annotations
  EntrezID, ID_mapping, GeneAnno,
  # Table related objects
  df, TB, Body_HTML, TableResHTML, TableResHTML.head, TableResHTML.body,
  # Ploting
  Plot, Plot.ls, FOVs.idx,
  # Seurat Objects
  xenium.obj.crop
)
gc()
save.image(file = "BreakPoint.RData")
```

## 02-04. E12.5 WT vs E13.5 WT{.tabset .tabset-fade}

Load data and set parameters.

```{r 02-04-0}
# load the Seurat object
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
# load environmental variables
load("GlobalVariables.RData")
load("Genes_IDConvert.RData")
# Parameters settings
surfix.exp <- "E13.5"                                   # Experimental group
surfix.ctr <- "E12.5"                                   # Control group
test.use <- "MAST"                                      # Test method
p.adj.thres <- 0.01                                     # p.adj threshold
log2FC.abs.thres <- 1                                   # |log2FC| threshold
Compare.name <- paste0("DEGs.Res.", test.use, ".", surfix.exp, ".vs.", surfix.ctr)
Compare.group <- "age"                                  # Compare group
Compare.Cell.Ls <- as.vector(unlist(CellKept[c(2,4)]))  # cells list to keep
Compare.FOVs <- FOVs[c(2,4)]                            # FOVs list to keep
FEA_ResNa <- paste0(surfix.exp, ".vs.", surfix.ctr)
```

### Analysis {.tabset .tabset-dropdown}

#### Do DE analysis

[Perform DE analysis using Model-based Analysis of Single-cell Transcriptomics (MAST) approach](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5)

DEGs were defined as p.adj < 0.01 and |avg_log2FC| > 1

```{r 02-04-1.1}
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls)
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
DEGs.Res <- data.frame()
for (i in levels(Data_compare$seurat_clusters)) {
  print(paste0(Sys.time(),": Processing Seurat_Cluster ", i))
  ident.1 <- paste0(i, "_", surfix.exp)
  ident.2 <- paste0(i, "_", surfix.ctr)
  if(
    or(
      sum(grepl(ident.1, Data_compare$celltype.stim))<3,
      sum(grepl(ident.2, Data_compare$celltype.stim))<3
    )
  ){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }else{
    temp <- FindMarkers(
      object = Data_compare, 
      ident.1 = ident.1, 
      ident.2 = ident.2,
      test.use = test.use
    )
  }
  if(nrow(temp)==0){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }
  temp <- cbind(temp, gene=rownames(temp), cluster = i)
  rownames(temp) <- paste0(rownames(temp),"_",i)
  DEGs.Res <- rbind(
    DEGs.Res, temp
  )
}
Term <- c(
  "Title",
  "GroupA", "GroupB", "Transformation of Expression", "Statistic Method",
  "avg_logFC","pct.1", "pct.2", "p_val_adj"
)
Value <- c(
  Compare.name,
  surfix.exp, surfix.ctr, DefaultAssay(Data_compare), test.use,
  "log fold-chage of the average expression between the two groups (GroupA/GroupB). Positive values indicate that the gene is more highly expressed in the GroupA",
  "The percentage of cells where the gene is detected in the GroupA",
  "The percentage of cells where the gene is detected in the GroupB",
  "Adjusted p-value, based on bonferroni correction using all genes in the dataset"
)
Comment <- c(
  "",
  "Experiment group, which is the numerator when calculating fold change",
  "Control group, which is the denominator when calculating fold change",
  "","","","","",""
)
length(Term) <- max(length(Term), length(Value), length(Comment))
DEGs.info <- data.frame(Term, Value, Comment)
```

Export the results, clean up, and save for the next step.

```{r 02-04-1.2}
xenium.obj.crop@misc[[Compare.name]] <- DEGs.Res
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
xlsx<-createWorkbook()
addWorksheet(xlsx,"DEGs_Res")
writeDataTable(xlsx,"DEGs_Res",DEGs.Res,startCol = 1,startRow = 1)
addWorksheet(xlsx,"Info")
writeDataTable(xlsx,"Info",DEGs.info,startCol = 1,startRow = 1)
saveWorkbook(xlsx, paste0(Compare.name, ".xlsx"), overwrite = TRUE)
rm(
  DEGs.Res, temp, ident.1, ident.2, CellTypes,
  Data_compare, DEGs.info, xlsx,
  Term, Value, Comment
)
gc()
save.image(file = "BreakPoint.RData")
```

#### Functional Enrichment Analysis

```{r 02-04-2.1}
xenium.obj.crop <- readRDS("xenium.obj.crop.rds")
temp <- ID_mapping
rownames(temp) <- make.unique(temp$GeneSymbol)
Data <- cbind(
  xenium.obj.crop@misc[[Compare.name]],
  temp[xenium.obj.crop@misc[[Compare.name]]$gene,]
)
Data$GeneSymbol <- Data$gene
Data <- Data[,-6]
WZY_RunFEA(
  log2FColNa = "avg_log2FC", ColChk = TRUE, Group.By = "cluster", Data = Data,
  tbOut = Data, WD = WD, FilePath = FEA_ResNa, ShinyProgress = FALSE,
  AnnoHub = "org.Mm.eg.db", Table1_rows_all = rownames(Data)[Data$p_val_adj<p.adj.thres&abs(Data$avg_log2FC)>log2FC.abs.thres],
  Par_UseFDR4FEA = FALSE, Par_FDR4FEA = 0.05, Par_FDR4KEG_OR = 0.05,
  Par_species.go.kegg = "mmu", descend = TRUE
)
```

Zip FEA output files for [Download](./E13.5.vs.E12.5_FEA_Res.zip)

```{bash 02-04-2.2, engine.opts='-l'}
zip -r ./E13.5.vs.E12.5_FEA_Res.zip ./E13.5.vs.E12.5_FEA_Res
```

Export the results, clean up, and save for the next step.

```{r 02-04-2.3}
rm(temp, Data)
gc()
```

### Spread Sheet

```{r 03-01.3.a, echo=FALSE, warning=FALSE, cache=FALSE, results='asis'}
TB <- xenium.obj.crop@misc[[Compare.name]]
TB <- TB[TB$p_val_adj<p.adj.thres &abs(TB$avg_log2FC)>log2FC.abs.thres,]
tableID<-"DEGs_1"
toggleID<-paste0(tableID,"FT")
colnm<-c(
  'p_val', 'avg_log2FC', 'pct.1', 'pct.2',
  'p_val_adj', 'gene', 'cluster'
)
TableResHTML.head<-hmakeTag(
  'th',colnm,
  style="font-weight: 900; border-bottom: 1px solid grey; border-top: 2px solid grey; text-align: center;"
  
)%>%
  paste(.,collapse="")%>%
  hmakeTag('tr',.)%>%hmakeTag('thead',.)
TableResHTML.body<-TB%>%
  hwrite(.,row.names=FALSE,col.names=FALSE,br=TRUE)%>%
  HTML()%>%read_html()%>%
  xml_find_all(.,".//tr")%>%as.character()%>%paste(.,collapse="")%>%
  hmakeTag('tbody',.)
TableResHTML<-hmakeTag('table',paste0(TableResHTML.head,TableResHTML.body))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//table")
ToggleBt<-hmakeTag('div', id=toggleID)
xml_set_attrs(TableResHTML[[1]], c("id"=tableID, "cellpadding"="0"))
JS_HTML_src<-hmakeTag('script', src='./HTML_source/tablefilter/tablefilter.js')
JS_HTML<-hmakeTag(
  'script', id='data-config',
  data = paste0(
    "
    var filtersConfig = {base_path: './HTML_source/tablefilter/', popup_filters: false, alternate_rows: true,
        msg_filter: 'Filtering...', rows_counter: true, btn_reset: true, status_bar: true,
        mark_active_columns: true, highlight_keywords: true, enable_default_theme: true,
        col_6: 'select',
        col_types: [
    ",
    paste(
      c('number', 'number', 'number', 'number', 'number',
        'string', 'number'), collapse="','"
    )%>%paste0("'",.,"'"),
    "
        ],
        extensions:[{name: 'sort' },
        {name: 'filtersVisibility', target_id: '",toggleID,"', btn_html: '<button>Toggle filters</button>'},
        {name: 'colsVisibility',
        text: 'Columns: ',tick_to_hide: false,enable_tick_all: true,
        headers_text: [
    ",
    paste(
      colnm, collapse="','"
    )%>%paste0("'",.,"'"),
    "
          ],
          at_start: []
        }],
        paging: {
          results_per_page: ['Records: ', [10, 25, 50, 100]]
        }
    };
    var tf = new TableFilter(",
    tableID%>%paste0("'",.,"'"),
    ", filtersConfig);
    tf.init();
    "
  )
)
Body_HTML<-hmakeTag('body',paste0(ToggleBt,TableResHTML,JS_HTML_src,JS_HTML))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//body")
as.character(Body_HTML)%>%HTML()
```

Click [HERE](./DEGs.Res.MAST.E13.5.vs.E12.5.xlsx) to download the above results as a .xlsx file.

### Distributions {.tabset .tabset-dropdown}

```{r}
TB <- xenium.obj.crop@misc[[Compare.name]]
df <- TB[TB$p_val_adj<p.adj.thres,]
df <- df[abs(df$avg_log2FC)>log2FC.abs.thres,]
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls) %>% suppressWarnings()
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
print("DEGs numbers per cluster")
print(table(df$cluster))
```

#### Seurat_cluster 0

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-0
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 1

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-1
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 2

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-2
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 3

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-3
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 4

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-4
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 5

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-5
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 6

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-6
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=12,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 7

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-7
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 8

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-8
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 9

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-9
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 10

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-10
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 11

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-11
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 12

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-12
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

### FEA Results {.tabset .tabset-dropdown}

Click [HERE](./E13.5.vs.E12.5_FEA_Res.zip) to download all FEA results. The MS .xlsx files of all FEA results are avalible inside the downloaded .zip file.

#### Seurat_Cluster 0

```{r, fig.width=24, fig.height=10}
i<-1
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 1

```{r, fig.width=24, fig.height=10}
i<-2
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 2

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 3

```{r, fig.width=24, fig.height=10}
i<-4
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 4

```{r, fig.width=24, fig.height=10}
i<-5
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 5

```{r, fig.width=24, fig.height=10}
i<-5
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 6

```{r, fig.width=24, fig.height=10}
i<-6
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 7

```{r, fig.width=24, fig.height=10}
i<-7
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 8

```{r, fig.width=24, fig.height=10}
i<-8
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 9

```{r, fig.width=24, fig.height=10}
i<-9
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 10

```{r, fig.width=24, fig.height=10}
i<-10
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 11

```{r, fig.width=24, fig.height=10}
i<-11
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 12

```{r, fig.width=24, fig.height=10}
i<-12
i2 <- 12
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

### Help {.tabset .tabset-dropdown}

#### About DEGs

Below is a demo to show how to check the Cell Markers results using our software:

<video width="1200" height="675" controls>
  <source src="CMarkers_Demo.mp4" type="video/mp4">
</video>

#### About FEA

Below is a demo to show how to check the FEA results using our software:

<video width="1200" height="675" controls>
  <source src="FEA_Res_Demo.mp4" type="video/mp4">
</video>

#### Miscellaneous

Save current results and cleanup for the next step

```{r 02-04-fin}
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
# Clean up
rm(
  Compare.name, Compare.group, Compare.Cell.Ls, Compare.FOVs,
  surfix.exp, surfix.ctr, test.use, p.adj.thres, log2FC.abs.thres, FEA_ResNa, 
  TB, temp, ident.1, ident.2, TableResHTML, TableResHTML.head, TableResHTML.body,
  FilePath, CellTypes, fig_svg,
  xenium.obj.crop
)
gc()
save.image(file = "BreakPoint.RData")
```

## 02-05. E12.5 WT vs E12.5 KO{.tabset .tabset-fade}

Load data and set parameters.

```{r 02-05-0}
# load the Seurat object
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
# load environmental variables
load("GlobalVariables.RData")
load("Genes_IDConvert.RData")
# Parameters settings
surfix.exp <- "PAX9-KO"                                 # Experimental group
surfix.ctr <- "WT"                                      # Control group
test.use <- "MAST"                                      # Test method
p.adj.thres <- 0.01                                     # p.adj threshold
log2FC.abs.thres <- 1                                   # |log2FC| threshold
Compare.name <- paste0("DEGs.Res.", test.use, ".E12.5_", surfix.exp, ".vs.", surfix.ctr)
Compare.group <- "condition"                            # Compare group
Compare.Cell.Ls <- as.vector(unlist(CellKept[c(1,2)]))  # cells list to keep
Compare.FOVs <- FOVs[c(1,2)]                            # FOVs list to keep
FEA_ResNa <- paste0("E12.5_", surfix.exp, ".vs.", surfix.ctr)
```

### Analysis {.tabset .tabset-dropdown}

#### Do DE analysis

[Perform DE analysis using Model-based Analysis of Single-cell Transcriptomics (MAST) approach](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5)

DEGs were defined as p.adj < 0.01 and |avg_log2FC| > 1

```{r 02-05-1.1}
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls)
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
DEGs.Res <- data.frame()
for (i in levels(Data_compare$seurat_clusters)) {
  print(paste0(Sys.time(),": Processing Seurat_Cluster ", i))
  ident.1 <- paste0(i, "_", surfix.exp)
  ident.2 <- paste0(i, "_", surfix.ctr)
  if(
    or(
      sum(grepl(ident.1, Data_compare$celltype.stim))<3,
      sum(grepl(ident.2, Data_compare$celltype.stim))<3
    )
  ){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }else{
    temp <- FindMarkers(
      object = Data_compare, 
      ident.1 = ident.1, 
      ident.2 = ident.2,
      test.use = test.use
    )
  }
  if(nrow(temp)==0){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }
  temp <- cbind(temp, gene=rownames(temp), cluster = i)
  rownames(temp) <- paste0(rownames(temp),"_",i)
  DEGs.Res <- rbind(
    DEGs.Res, temp
  )
}
Term <- c(
  "Title",
  "GroupA", "GroupB", "Transformation of Expression", "Statistic Method",
  "avg_logFC","pct.1", "pct.2", "p_val_adj"
)
Value <- c(
  Compare.name,
  surfix.exp, surfix.ctr, DefaultAssay(Data_compare), test.use,
  "log fold-chage of the average expression between the two groups (GroupA/GroupB). Positive values indicate that the gene is more highly expressed in the GroupA",
  "The percentage of cells where the gene is detected in the GroupA",
  "The percentage of cells where the gene is detected in the GroupB",
  "Adjusted p-value, based on bonferroni correction using all genes in the dataset"
)
Comment <- c(
  "",
  "Experiment group, which is the numerator when calculating fold change",
  "Control group, which is the denominator when calculating fold change",
  "","","","","",""
)
length(Term) <- max(length(Term), length(Value), length(Comment))
DEGs.info <- data.frame(Term, Value, Comment)
```

Export the results, clean up, and save for the next step.

```{r 02-05-1.2}
xenium.obj.crop@misc[[Compare.name]] <- DEGs.Res
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
xlsx<-createWorkbook()
addWorksheet(xlsx,"DEGs_Res")
writeDataTable(xlsx,"DEGs_Res",DEGs.Res,startCol = 1,startRow = 1)
addWorksheet(xlsx,"Info")
writeDataTable(xlsx,"Info",DEGs.info,startCol = 1,startRow = 1)
saveWorkbook(xlsx, paste0(Compare.name, ".xlsx"), overwrite = TRUE)
rm(
  DEGs.Res, temp, ident.1, ident.2, CellTypes,
  Data_compare, DEGs.info, xlsx,
  Term, Value, Comment
)
gc()
save.image(file = "BreakPoint.RData")
```

#### Functional Enrichment Analysis

```{r 02-05-2.1}
xenium.obj.crop <- readRDS("xenium.obj.crop.rds")
temp <- ID_mapping
rownames(temp) <- make.unique(temp$GeneSymbol)
Data <- cbind(
  xenium.obj.crop@misc[[Compare.name]],
  temp[xenium.obj.crop@misc[[Compare.name]]$gene,]
)
Data$GeneSymbol <- Data$gene
Data <- Data[,-6]
WZY_RunFEA(
  log2FColNa = "avg_log2FC", ColChk = TRUE, Group.By = "cluster", Data = Data,
  tbOut = Data, WD = WD, FilePath = FEA_ResNa, ShinyProgress = FALSE,
  AnnoHub = "org.Mm.eg.db", Table1_rows_all = rownames(Data)[Data$p_val_adj<p.adj.thres&abs(Data$avg_log2FC)>log2FC.abs.thres],
  Par_UseFDR4FEA = FALSE, Par_FDR4FEA = 0.05, Par_FDR4KEG_OR = 0.05,
  Par_species.go.kegg = "mmu", descend = TRUE
)
```

Zip FEA output files for [Download](./E12.5_PAX9-KO.vs.WT_FEA_Res.zip)

```{bash 02-05-2.2, engine.opts='-l'}
zip -r ./E12.5_PAX9-KO.vs.WT_FEA_Res.zip ./E12.5_PAX9-KO.vs.WT_FEA_Res
```

Export the results, clean up, and save for the next step.

```{r 02-05-2.3}
rm(temp, Data)
gc()
```

### Spread Sheet

```{r 02-05-3, echo=FALSE, warning=FALSE, cache=FALSE, results='asis'}
TB <- xenium.obj.crop@misc[[Compare.name]]
TB <- TB[TB$p_val_adj<p.adj.thres &abs(TB$avg_log2FC)>log2FC.abs.thres,]
tableID<-"DEGs_2"
toggleID<-paste0(tableID,"FT")
colnm<-c(
  'p_val', 'avg_log2FC', 'pct.1', 'pct.2',
  'p_val_adj', 'gene', 'cluster'
)
TableResHTML.head<-hmakeTag(
  'th',colnm,
  style="font-weight: 900; border-bottom: 1px solid grey; border-top: 2px solid grey; text-align: center;"
  
)%>%
  paste(.,collapse="")%>%
  hmakeTag('tr',.)%>%hmakeTag('thead',.)
TableResHTML.body<-TB%>%
  hwrite(.,row.names=FALSE,col.names=FALSE,br=TRUE)%>%
  HTML()%>%read_html()%>%
  xml_find_all(.,".//tr")%>%as.character()%>%paste(.,collapse="")%>%
  hmakeTag('tbody',.)
TableResHTML<-hmakeTag('table',paste0(TableResHTML.head,TableResHTML.body))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//table")
ToggleBt<-hmakeTag('div', id=toggleID)
xml_set_attrs(TableResHTML[[1]], c("id"=tableID, "cellpadding"="0"))
JS_HTML_src<-hmakeTag('script', src='./HTML_source/tablefilter/tablefilter.js')
JS_HTML<-hmakeTag(
  'script', id='data-config',
  data = paste0(
    "
    var filtersConfig = {base_path: './HTML_source/tablefilter/', popup_filters: false, alternate_rows: true,
        msg_filter: 'Filtering...', rows_counter: true, btn_reset: true, status_bar: true,
        mark_active_columns: true, highlight_keywords: true, enable_default_theme: true,
        col_6: 'select',
        col_types: [
    ",
    paste(
      c('number', 'number', 'number', 'number', 'number',
        'string', 'number'), collapse="','"
    )%>%paste0("'",.,"'"),
    "
        ],
        extensions:[{name: 'sort' },
        {name: 'filtersVisibility', target_id: '",toggleID,"', btn_html: '<button>Toggle filters</button>'},
        {name: 'colsVisibility',
        text: 'Columns: ',tick_to_hide: false,enable_tick_all: true,
        headers_text: [
    ",
    paste(
      colnm, collapse="','"
    )%>%paste0("'",.,"'"),
    "
          ],
          at_start: []
        }],
        paging: {
          results_per_page: ['Records: ', [10, 25, 50, 100]]
        }
    };
    var tf = new TableFilter(",
    tableID%>%paste0("'",.,"'"),
    ", filtersConfig);
    tf.init();
    "
  )
)
Body_HTML<-hmakeTag('body',paste0(ToggleBt,TableResHTML,JS_HTML_src,JS_HTML))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//body")
as.character(Body_HTML)%>%HTML()
```

Click [HERE](./DEGs.Res.MAST.E12.5_PAX9-KO.vs.WT.xlsx) to download the above results as a .xlsx file.

### Distributions {.tabset .tabset-dropdown}

```{r 02-05-4}
TB <- xenium.obj.crop@misc[[Compare.name]]
df <- TB[TB$p_val_adj<p.adj.thres,]
df <- df[abs(df$avg_log2FC)>log2FC.abs.thres,]
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls) %>% suppressWarnings()
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
print("DEGs numbers per cluster")
print(table(df$cluster))
```

#### Seurat_cluster 0

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-0
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 1

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-1
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 2

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-2
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 3

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-3
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 4

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-4
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 5

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-5
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 6

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-6
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 7

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-7
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 8

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-8
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 9

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-9
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 10

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-10
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 11

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-11
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 12

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-12
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

### FEA Results {.tabset .tabset-dropdown}

Click [HERE](./E12.5_PAX9-KO.vs.WT_FEA_Res.zip) to download all FEA results. The MS .xlsx files of all FEA results are avalible inside the downloaded .zip file.

#### Seurat_Cluster 0

```{r, fig.width=24, fig.height=10}
i<-1
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 1

```{r, fig.width=24, fig.height=10}
i<-1
i2 <- i
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 2

```{r, fig.width=24, fig.height=10}
i<-2
i2 <- i
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 3

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- 3
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 4

```{r, fig.width=24, fig.height=10}
i<-4
i2 <- 4
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 5

```{r, fig.width=24, fig.height=10}
i<-5
i2 <- 5
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 6

```{r, fig.width=24, fig.height=10}
i<-6
i2 <- 5
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 7

```{r, fig.width=24, fig.height=10}
i<-7
i2 <- 7
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 8

```{r, fig.width=24, fig.height=10}
i<-8
i2 <- 8
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 9

```{r, fig.width=24, fig.height=10}
i<-9
i2 <- 9
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 10

```{r, fig.width=24, fig.height=10}
i<-10
i2 <- 10
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 11

```{r, fig.width=24, fig.height=10}
i<-11
i2 <- 11
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 12

```{r, fig.width=24, fig.height=10}
i<-12
i2 <- 12
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

### Help {.tabset .tabset-dropdown}

#### About DEGs

Below is a demo to show how to check the Cell Markers results using our software:

<video width="1200" height="675" controls>
  <source src="CMarkers_Demo.mp4" type="video/mp4">
</video>

#### About FEA

Below is a demo to show how to check the FEA results using our software:

<video width="1200" height="675" controls>
  <source src="FEA_Res_Demo.mp4" type="video/mp4">
</video>

#### Miscellaneous

Save current results and cleanup for the next step

```{r 02-05-fin}
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
# Clean up
rm(
  Compare.name, Compare.group, Compare.Cell.Ls, Compare.FOVs,
  surfix.exp, surfix.ctr, test.use, p.adj.thres, log2FC.abs.thres, FEA_ResNa, 
  TB, temp, ident.1, ident.2, TableResHTML, TableResHTML.head, TableResHTML.body,
  FilePath, CellTypes, fig_svg,
  xenium.obj.crop
)
gc()
save.image(file = "BreakPoint.RData")
```

## 02-06. E13.5 WT vs E13.5 KO{.tabset .tabset-fade}

Load data and set parameters.

```{r 02-06-0}
# load the Seurat object
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
# load environmental variables
load("GlobalVariables.RData")
load("Genes_IDConvert.RData")
# Parameters settings
surfix.exp <- "PAX9-KO"                                 # Experimental group
surfix.ctr <- "WT"                                      # Control group
test.use <- "MAST"                                      # Test method
p.adj.thres <- 0.01                                     # p.adj threshold
log2FC.abs.thres <- 1                                   # |log2FC| threshold
Compare.name <- paste0("DEGs.Res.", test.use, ".E13.5_", surfix.exp, ".vs.", surfix.ctr)
Compare.group <- "condition"                            # Compare group
Compare.Cell.Ls <- as.vector(unlist(CellKept[c(3,4)]))  # cells list to keep
Compare.FOVs <- FOVs[c(3,4)]                            # FOVs list to keep
FEA_ResNa <- paste0("E13.5_", surfix.exp, ".vs.", surfix.ctr)
```

### Analysis {.tabset .tabset-dropdown}

#### Do DE analysis

[Perform DE analysis using Model-based Analysis of Single-cell Transcriptomics (MAST) approach](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5)

DEGs were defined as p.adj < 0.01 and |avg_log2FC| > 1

```{r 02-06-1.1}
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls)
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
DEGs.Res <- data.frame()
for (i in levels(Data_compare$seurat_clusters)) {
  print(paste0(Sys.time(),": Processing Seurat_Cluster ", i))
  ident.1 <- paste0(i, "_", surfix.exp)
  ident.2 <- paste0(i, "_", surfix.ctr)
  if(
    or(
      sum(grepl(ident.1, Data_compare$celltype.stim))<3,
      sum(grepl(ident.2, Data_compare$celltype.stim))<3
    )
  ){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }else{
    temp <- FindMarkers(
      object = Data_compare, 
      ident.1 = ident.1, 
      ident.2 = ident.2,
      test.use = test.use
    )
  }
  if(nrow(temp)==0){
    temp <- data.frame(
      p_val = 1,
      avg_log2FC = 0,
      pct.1 = sum(grepl(ident.1, Data_compare$celltype.stim)),
      pct.2 = sum(grepl(ident.2, Data_compare$celltype.stim)),
      p_val_adj = 1
    )
    rownames(temp) <- "NoDEGs"
  }
  temp <- cbind(temp, gene=rownames(temp), cluster = i)
  rownames(temp) <- paste0(rownames(temp),"_",i)
  DEGs.Res <- rbind(
    DEGs.Res, temp
  )
}
Term <- c(
  "Title",
  "GroupA", "GroupB", "Transformation of Expression", "Statistic Method",
  "avg_logFC","pct.1", "pct.2", "p_val_adj"
)
Value <- c(
  Compare.name,
  surfix.exp, surfix.ctr, DefaultAssay(Data_compare), test.use,
  "log fold-chage of the average expression between the two groups (GroupA/GroupB). Positive values indicate that the gene is more highly expressed in the GroupA",
  "The percentage of cells where the gene is detected in the GroupA",
  "The percentage of cells where the gene is detected in the GroupB",
  "Adjusted p-value, based on bonferroni correction using all genes in the dataset"
)
Comment <- c(
  "",
  "Experiment group, which is the numerator when calculating fold change",
  "Control group, which is the denominator when calculating fold change",
  "","","","","",""
)
length(Term) <- max(length(Term), length(Value), length(Comment))
DEGs.info <- data.frame(Term, Value, Comment)
```

Export the results, clean up, and save for the next step.

```{r 02-06-1.2}
xenium.obj.crop@misc[[Compare.name]] <- DEGs.Res
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
xlsx<-createWorkbook()
addWorksheet(xlsx,"DEGs_Res")
writeDataTable(xlsx,"DEGs_Res",DEGs.Res,startCol = 1,startRow = 1)
addWorksheet(xlsx,"Info")
writeDataTable(xlsx,"Info",DEGs.info,startCol = 1,startRow = 1)
saveWorkbook(xlsx, paste0(Compare.name, ".xlsx"), overwrite = TRUE)
rm(
  DEGs.Res, temp, ident.1, ident.2, CellTypes,
  Data_compare, DEGs.info, xlsx,
  Term, Value, Comment
)
gc()
save.image(file = "BreakPoint.RData")
```

#### Functional Enrichment Analysis

```{r 02-06-2.1}
xenium.obj.crop <- readRDS("xenium.obj.crop.rds")
temp <- ID_mapping
rownames(temp) <- make.unique(temp$GeneSymbol)
Data <- cbind(
  xenium.obj.crop@misc[[Compare.name]],
  temp[xenium.obj.crop@misc[[Compare.name]]$gene,]
)
Data$GeneSymbol <- Data$gene
Data <- Data[,-6]
WZY_RunFEA(
  log2FColNa = "avg_log2FC", ColChk = TRUE, Group.By = "cluster", Data = Data,
  tbOut = Data, WD = WD, FilePath = FEA_ResNa, ShinyProgress = FALSE,
  AnnoHub = "org.Mm.eg.db", Table1_rows_all = rownames(Data)[Data$p_val_adj<p.adj.thres&abs(Data$avg_log2FC)>log2FC.abs.thres],
  Par_UseFDR4FEA = FALSE, Par_FDR4FEA = 0.05, Par_FDR4KEG_OR = 0.05,
  Par_species.go.kegg = "mmu", descend = TRUE
)
```

Zip FEA output files for [Download](./E13.5_PAX9-KO.vs.WT_FEA_Res.zip)

```{bash 02-06-2.2, engine.opts='-l'}
zip -r ./E13.5_PAX9-KO.vs.WT_FEA_Res.zip ./E13.5_PAX9-KO.vs.WT_FEA_Res
```

Export the results, clean up, and save for the next step.

```{r 02-06-2.3}
rm(temp, Data)
gc()
```

### Spread Sheet

```{r 02-06-3, echo=FALSE, warning=FALSE, cache=FALSE, results='asis'}
TB <- xenium.obj.crop@misc[[Compare.name]]
TB <- TB[TB$p_val_adj<p.adj.thres &abs(TB$avg_log2FC)>log2FC.abs.thres,]
tableID<-"DEGs_3"
toggleID<-paste0(tableID,"FT")
colnm<-c(
  'p_val', 'avg_log2FC', 'pct.1', 'pct.2',
  'p_val_adj', 'gene', 'cluster'
)
TableResHTML.head<-hmakeTag(
  'th',colnm,
  style="font-weight: 900; border-bottom: 1px solid grey; border-top: 2px solid grey; text-align: center;"
  
)%>%
  paste(.,collapse="")%>%
  hmakeTag('tr',.)%>%hmakeTag('thead',.)
TableResHTML.body<-TB%>%
  hwrite(.,row.names=FALSE,col.names=FALSE,br=TRUE)%>%
  HTML()%>%read_html()%>%
  xml_find_all(.,".//tr")%>%as.character()%>%paste(.,collapse="")%>%
  hmakeTag('tbody',.)
TableResHTML<-hmakeTag('table',paste0(TableResHTML.head,TableResHTML.body))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//table")
ToggleBt<-hmakeTag('div', id=toggleID)
xml_set_attrs(TableResHTML[[1]], c("id"=tableID, "cellpadding"="0"))
JS_HTML_src<-hmakeTag('script', src='./HTML_source/tablefilter/tablefilter.js')
JS_HTML<-hmakeTag(
  'script', id='data-config',
  data = paste0(
    "
    var filtersConfig = {base_path: './HTML_source/tablefilter/', popup_filters: false, alternate_rows: true,
        msg_filter: 'Filtering...', rows_counter: true, btn_reset: true, status_bar: true,
        mark_active_columns: true, highlight_keywords: true, enable_default_theme: true,
        col_6: 'select',
        col_types: [
    ",
    paste(
      c('number', 'number', 'number', 'number', 'number',
        'string', 'number'), collapse="','"
    )%>%paste0("'",.,"'"),
    "
        ],
        extensions:[{name: 'sort' },
        {name: 'filtersVisibility', target_id: '",toggleID,"', btn_html: '<button>Toggle filters</button>'},
        {name: 'colsVisibility',
        text: 'Columns: ',tick_to_hide: false,enable_tick_all: true,
        headers_text: [
    ",
    paste(
      colnm, collapse="','"
    )%>%paste0("'",.,"'"),
    "
          ],
          at_start: []
        }],
        paging: {
          results_per_page: ['Records: ', [10, 25, 50, 100]]
        }
    };
    var tf = new TableFilter(",
    tableID%>%paste0("'",.,"'"),
    ", filtersConfig);
    tf.init();
    "
  )
)
Body_HTML<-hmakeTag('body',paste0(ToggleBt,TableResHTML,JS_HTML_src,JS_HTML))%>%
  HTML()%>%read_html()%>%xml_find_all(.,".//body")
as.character(Body_HTML)%>%HTML()
```

Click [HERE](./DEGs.Res.MAST.E13.5.vs.E12.5.xlsx) to download the above results as a .xlsx file.

### Distributions {.tabset .tabset-dropdown}

```{r 02-06-4}
TB <- xenium.obj.crop@misc[[Compare.name]]
df <- TB[TB$p_val_adj<p.adj.thres,]
df <- df[abs(df$avg_log2FC)>log2FC.abs.thres,]
Data_compare <- xenium.obj.crop
Data_compare <- subset(Data_compare, cells = Compare.Cell.Ls) %>% suppressWarnings()
Data_compare$celltype.stim <- paste(Data_compare$seurat_clusters, str_replace_all(Data_compare[[Compare.group]][[1]],"_","-"), sep = "_")
Idents(Data_compare) <- "celltype.stim"
CellTypes <-  levels(Data_compare)
print("DEGs numbers per cluster")
print(table(df$cluster))
```

#### Seurat_cluster 0

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-0
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 1

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-1
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 2

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-2
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 3

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-3
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 4

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-4
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=28,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 5

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-5
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=20,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 6

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-6
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 7

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-7
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 8

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-8
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 9

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-9
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 10

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-10
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 11

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-11
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

#### Seurat_cluster 12

```{r,fig.width=16,fig.height=8,message=FALSE,warning=FALSE}
i<-12
temp <- df[df$cluster==i,]
temp <- temp[order(abs(temp$avg_log2FC), decreasing = TRUE),]
if(nrow(temp)<6){nu <- nrow(temp)}else{nu<-6}
Top6 <- temp$gene[1:nu]
ident.1 <- paste0(i, "_", surfix.exp)
ident.2 <- paste0(i, "_", surfix.ctr)
if(!nrow(temp)){
  print(paste0("There are no significant DEGs in the seurat_cluster ", i, "."))
  print(paste0("This may due to the insufficient cells numbers in either groups."))
  print(paste0("Please use the Loupe Browser Software to check the cells numbers."))
}else{
  VlnPlot(Data_compare, features = Top6, idents = c(ident.2, ident.1), group.by = Compare.group, assay="SCT", ncol = length(Top6), add.noise = TRUE)
}
```

```{r,fig.width=16,fig.height=4,message=FALSE,warning=FALSE}
if(!nrow(temp)){
  do.call("grid.arrange", c(
    WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=TRUE), ncol=2,
    top = paste0("There are no significant DEGs in the seurat_cluster ", i, ".")
  ))
}else{
  do.call("grid.arrange", c(WZY_FeatureBlukPlot(i=i, Data_temp=Data_compare, geneLs = Top6, fovs = Compare.FOVs, OnlyCluster=FALSE), ncol=2))
}
```

### FEA Results {.tabset .tabset-dropdown}

Click [HERE](./E13.5.vs.E12.5_FEA_Res.zip) to download all FEA results. The MS .xlsx files of all FEA results are avalible inside the downloaded .zip file.

#### Seurat_Cluster 0

```{r, fig.width=24, fig.height=10}
i<-1
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 1

```{r, fig.width=24, fig.height=10}
i<-2
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 2

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- i-1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 3

```{r, fig.width=24, fig.height=10}
i<-3
i2 <- 3
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 4

```{r, fig.width=24, fig.height=10}
i<-4
i2 <- 4
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 5

```{r, fig.width=24, fig.height=10}
i<-5
i2 <- 5
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 6

```{r, fig.width=24, fig.height=10}
i<-6
i2 <- i
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 7

```{r, fig.width=24, fig.height=10}
i<-7
i2 <- i
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 8

```{r, fig.width=24, fig.height=10}
i<-7
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 9

```{r, fig.width=24, fig.height=10}
i<-8
i2 <- i+1
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 10

```{r, fig.width=24, fig.height=10}
i<-6
i2 <- 10
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_KEGG.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

#### Seurat_Cluster 11

```{r, fig.width=24, fig.height=10}
i<-11
i2 <- 11
FilePath <- file.path(
  ".",paste0(FEA_ResNa, "_FEA_Res"),"01.PlotOutput",
  paste0("FEA_Plot_",sprintf("%02d", i),"_",i2), "Over_Represent", "All_GO_BP.svg"
)
if(!file.exists(FilePath)){
  print("There is no significant enriched functional terms.")
}else{
  fig_svg<-magick::image_read_svg(FilePath)
  par(mar = c(0,0,0,0), xpd=NA)
  plot(fig_svg)
}
```

### Help {.tabset .tabset-dropdown}

#### About DEGs

Below is a demo to show how to check the Cell Markers results using our software:

<video width="1200" height="675" controls>
  <source src="CMarkers_Demo.mp4" type="video/mp4">
</video>

#### About FEA

Below is a demo to show how to check the FEA results using our software:

<video width="1200" height="675" controls>
  <source src="FEA_Res_Demo.mp4" type="video/mp4">
</video>

#### Miscellaneous

Save current results and cleanup for the next step

```{r 02-06-fin}
# Save the Seurat object
saveRDS(xenium.obj.crop, file = "xenium.obj.crop.rds")
# Clean up
rm(
  Compare.name, Compare.group, Compare.Cell.Ls, Compare.FOVs,
  surfix.exp, surfix.ctr, test.use, p.adj.thres, log2FC.abs.thres, FEA_ResNa, 
  TB, temp, ident.1, ident.2, TableResHTML, TableResHTML.head, TableResHTML.body,
  FilePath, CellTypes, fig_svg,
  xenium.obj.crop
)
gc()
save.image(file = "BreakPoint.RData")
```

# Supplementary

```{r final}
# Export Seurat object to cloupe file
xenium.obj.crop <- readRDS(file = "xenium.obj.crop.rds")
# Gene Expression RNA assay
assay <- xenium.obj.crop[["Xenium"]]
# get counts matrix from either the old or newer formats of assay
counts <- counts_matrix_from_assay(assay)
# get clusters from the Seurat object
clusters <- select_clusters(xenium.obj.crop)
# get projections from the Seurat object
projections <- select_projections(xenium.obj.crop)
# get morphologies from the Seurat object
morph <- xenium.obj.crop@images[["fov"]]@boundaries[["centroids"]]@coords %>% as.matrix()
rownames(morph) <- xenium.obj.crop@images[["fov"]]@boundaries[["centroids"]]@cells
# add the morphologies to the projections
projections <- c(projections, morph = list(morph))
# get the barcodes from the counts matrix
barcodes <- colnames(counts)
barcode_count <- length(barcodes)
# validate the data
loupeR:::validate_count_mat(counts)
loupeR:::validate_clusters(clusters, barcode_count)
loupeR:::validate_projections(projections, barcode_count)
# convert the count matrix, clusters, and projections into a Loupe file
if(!dir.exists("./00.WZY_Res")){
  dir.create("./00.WZY_Res")
}
create_loupe(
  counts,
  clusters = clusters,
  projections = projections,
  force = TRUE,
  output_dir = "./00.WZY_Res",
  output_name = "WZY_Res.crop",
  feature_ids = NULL,
  executable_path = NULL,
  seurat_obj_version = NULL
)
rm(
  assay, counts, clusters, projections, morph,
  barcodes, barcode_count, xenium.obj.crop
)
gc()
# Show environment
sessionInfo()
```

<script src="http://kit.fontawesome.com/46057cf47a.js" crossorigin="anonymous"></script>
<script src="./HTML_source/script.js"></script>